{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "signed-district",
   "metadata": {},
   "source": [
    "# Resource conservation in genetic code, using mean codon abundances"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "parliamentary-impossible",
   "metadata": {},
   "source": [
    "This notebook runs thes same analysis as **Resource_conservation_in_genetic_code.ipynb**, only using the same treatment of codon abundances as was used by Shenhav and Zeevi, Science, 2020, i.e., setting frequency of codon _v_ to the median of frequencies of all codons that encode the same amino acid as _v_ in the standard genetic code.\n",
    "\n",
    "Requirements: The code is written in **Python 3**. The following packages need to be installed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "aboriginal-bumper",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas\n",
    "import matplotlib.pyplot as plt\n",
    "from statistics import median"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "double-crossing",
   "metadata": {},
   "source": [
    "## Helper functions and global variables"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "welsh-happiness",
   "metadata": {},
   "source": [
    "### Transitions and transversions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "outside-riding",
   "metadata": {},
   "outputs": [],
   "source": [
    "transitions = {'A':'G', 'G':'A', 'C':'T', 'T':'C'}\n",
    "transversions = {'A':['C', 'T'], 'G':['C', 'T'], 'C':['A', 'G'], 'T':['A', 'G']}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "funny-terrace",
   "metadata": {},
   "source": [
    "Function that generates all neighbouring codons of a codon:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "civilian-congo",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gen_neighbouring_codons(c):\n",
    "    res = []\n",
    "    for i in [0,1,2]:\n",
    "        for b in ['A', 'C', 'G', 'T']:\n",
    "            if c[i]!=b:\n",
    "                res.append(c[:i] + b + c[(i+1):])\n",
    "    return res"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "addressed-zimbabwe",
   "metadata": {},
   "source": [
    "Function that determines type of substitution needed to convert codon *old* to codon *new*:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "specialized-circumstances",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_mutation_type(old, new):\n",
    "    for i in range(3):\n",
    "        if old[i]!=new[i]:\n",
    "            return old[i]+new[i]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "outdoor-latter",
   "metadata": {},
   "source": [
    "### Amino acid properties"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "refined-allowance",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S</th>\n",
       "      <th>R</th>\n",
       "      <th>L</th>\n",
       "      <th>P</th>\n",
       "      <th>T</th>\n",
       "      <th>A</th>\n",
       "      <th>V</th>\n",
       "      <th>G</th>\n",
       "      <th>I</th>\n",
       "      <th>F</th>\n",
       "      <th>Y</th>\n",
       "      <th>C</th>\n",
       "      <th>H</th>\n",
       "      <th>Q</th>\n",
       "      <th>N</th>\n",
       "      <th>K</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>M</th>\n",
       "      <th>W</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>0</td>\n",
       "      <td>110</td>\n",
       "      <td>145</td>\n",
       "      <td>74</td>\n",
       "      <td>58</td>\n",
       "      <td>99</td>\n",
       "      <td>124</td>\n",
       "      <td>56</td>\n",
       "      <td>142</td>\n",
       "      <td>155</td>\n",
       "      <td>144</td>\n",
       "      <td>112</td>\n",
       "      <td>89</td>\n",
       "      <td>68</td>\n",
       "      <td>46</td>\n",
       "      <td>121</td>\n",
       "      <td>65</td>\n",
       "      <td>80</td>\n",
       "      <td>135</td>\n",
       "      <td>177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R</th>\n",
       "      <td>110</td>\n",
       "      <td>0</td>\n",
       "      <td>102</td>\n",
       "      <td>103</td>\n",
       "      <td>71</td>\n",
       "      <td>112</td>\n",
       "      <td>96</td>\n",
       "      <td>125</td>\n",
       "      <td>97</td>\n",
       "      <td>97</td>\n",
       "      <td>77</td>\n",
       "      <td>180</td>\n",
       "      <td>29</td>\n",
       "      <td>43</td>\n",
       "      <td>86</td>\n",
       "      <td>26</td>\n",
       "      <td>96</td>\n",
       "      <td>54</td>\n",
       "      <td>91</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L</th>\n",
       "      <td>145</td>\n",
       "      <td>102</td>\n",
       "      <td>0</td>\n",
       "      <td>98</td>\n",
       "      <td>92</td>\n",
       "      <td>96</td>\n",
       "      <td>32</td>\n",
       "      <td>138</td>\n",
       "      <td>5</td>\n",
       "      <td>22</td>\n",
       "      <td>36</td>\n",
       "      <td>198</td>\n",
       "      <td>99</td>\n",
       "      <td>113</td>\n",
       "      <td>153</td>\n",
       "      <td>107</td>\n",
       "      <td>172</td>\n",
       "      <td>138</td>\n",
       "      <td>15</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P</th>\n",
       "      <td>74</td>\n",
       "      <td>103</td>\n",
       "      <td>98</td>\n",
       "      <td>0</td>\n",
       "      <td>38</td>\n",
       "      <td>27</td>\n",
       "      <td>68</td>\n",
       "      <td>42</td>\n",
       "      <td>95</td>\n",
       "      <td>114</td>\n",
       "      <td>110</td>\n",
       "      <td>169</td>\n",
       "      <td>77</td>\n",
       "      <td>76</td>\n",
       "      <td>91</td>\n",
       "      <td>103</td>\n",
       "      <td>108</td>\n",
       "      <td>93</td>\n",
       "      <td>87</td>\n",
       "      <td>147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>58</td>\n",
       "      <td>71</td>\n",
       "      <td>92</td>\n",
       "      <td>38</td>\n",
       "      <td>0</td>\n",
       "      <td>58</td>\n",
       "      <td>69</td>\n",
       "      <td>59</td>\n",
       "      <td>89</td>\n",
       "      <td>103</td>\n",
       "      <td>92</td>\n",
       "      <td>149</td>\n",
       "      <td>47</td>\n",
       "      <td>42</td>\n",
       "      <td>65</td>\n",
       "      <td>78</td>\n",
       "      <td>85</td>\n",
       "      <td>65</td>\n",
       "      <td>81</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>99</td>\n",
       "      <td>112</td>\n",
       "      <td>96</td>\n",
       "      <td>27</td>\n",
       "      <td>58</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>60</td>\n",
       "      <td>94</td>\n",
       "      <td>113</td>\n",
       "      <td>112</td>\n",
       "      <td>195</td>\n",
       "      <td>86</td>\n",
       "      <td>91</td>\n",
       "      <td>111</td>\n",
       "      <td>106</td>\n",
       "      <td>126</td>\n",
       "      <td>107</td>\n",
       "      <td>84</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V</th>\n",
       "      <td>124</td>\n",
       "      <td>96</td>\n",
       "      <td>32</td>\n",
       "      <td>68</td>\n",
       "      <td>69</td>\n",
       "      <td>64</td>\n",
       "      <td>0</td>\n",
       "      <td>109</td>\n",
       "      <td>29</td>\n",
       "      <td>50</td>\n",
       "      <td>55</td>\n",
       "      <td>192</td>\n",
       "      <td>84</td>\n",
       "      <td>96</td>\n",
       "      <td>133</td>\n",
       "      <td>97</td>\n",
       "      <td>152</td>\n",
       "      <td>121</td>\n",
       "      <td>21</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>56</td>\n",
       "      <td>125</td>\n",
       "      <td>138</td>\n",
       "      <td>42</td>\n",
       "      <td>59</td>\n",
       "      <td>60</td>\n",
       "      <td>109</td>\n",
       "      <td>0</td>\n",
       "      <td>135</td>\n",
       "      <td>153</td>\n",
       "      <td>147</td>\n",
       "      <td>159</td>\n",
       "      <td>98</td>\n",
       "      <td>87</td>\n",
       "      <td>80</td>\n",
       "      <td>127</td>\n",
       "      <td>94</td>\n",
       "      <td>98</td>\n",
       "      <td>127</td>\n",
       "      <td>184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I</th>\n",
       "      <td>142</td>\n",
       "      <td>97</td>\n",
       "      <td>5</td>\n",
       "      <td>95</td>\n",
       "      <td>89</td>\n",
       "      <td>94</td>\n",
       "      <td>29</td>\n",
       "      <td>135</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>33</td>\n",
       "      <td>198</td>\n",
       "      <td>94</td>\n",
       "      <td>109</td>\n",
       "      <td>149</td>\n",
       "      <td>102</td>\n",
       "      <td>168</td>\n",
       "      <td>134</td>\n",
       "      <td>10</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>155</td>\n",
       "      <td>97</td>\n",
       "      <td>22</td>\n",
       "      <td>114</td>\n",
       "      <td>103</td>\n",
       "      <td>113</td>\n",
       "      <td>50</td>\n",
       "      <td>153</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>205</td>\n",
       "      <td>100</td>\n",
       "      <td>116</td>\n",
       "      <td>158</td>\n",
       "      <td>102</td>\n",
       "      <td>177</td>\n",
       "      <td>140</td>\n",
       "      <td>28</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <td>144</td>\n",
       "      <td>77</td>\n",
       "      <td>36</td>\n",
       "      <td>110</td>\n",
       "      <td>92</td>\n",
       "      <td>112</td>\n",
       "      <td>55</td>\n",
       "      <td>147</td>\n",
       "      <td>33</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>194</td>\n",
       "      <td>83</td>\n",
       "      <td>99</td>\n",
       "      <td>143</td>\n",
       "      <td>85</td>\n",
       "      <td>160</td>\n",
       "      <td>122</td>\n",
       "      <td>36</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>112</td>\n",
       "      <td>180</td>\n",
       "      <td>198</td>\n",
       "      <td>169</td>\n",
       "      <td>149</td>\n",
       "      <td>195</td>\n",
       "      <td>192</td>\n",
       "      <td>159</td>\n",
       "      <td>198</td>\n",
       "      <td>205</td>\n",
       "      <td>194</td>\n",
       "      <td>0</td>\n",
       "      <td>174</td>\n",
       "      <td>154</td>\n",
       "      <td>139</td>\n",
       "      <td>202</td>\n",
       "      <td>154</td>\n",
       "      <td>170</td>\n",
       "      <td>196</td>\n",
       "      <td>215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>89</td>\n",
       "      <td>29</td>\n",
       "      <td>99</td>\n",
       "      <td>77</td>\n",
       "      <td>47</td>\n",
       "      <td>86</td>\n",
       "      <td>84</td>\n",
       "      <td>98</td>\n",
       "      <td>94</td>\n",
       "      <td>100</td>\n",
       "      <td>83</td>\n",
       "      <td>174</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>68</td>\n",
       "      <td>32</td>\n",
       "      <td>81</td>\n",
       "      <td>40</td>\n",
       "      <td>87</td>\n",
       "      <td>115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q</th>\n",
       "      <td>68</td>\n",
       "      <td>43</td>\n",
       "      <td>113</td>\n",
       "      <td>76</td>\n",
       "      <td>42</td>\n",
       "      <td>91</td>\n",
       "      <td>96</td>\n",
       "      <td>87</td>\n",
       "      <td>109</td>\n",
       "      <td>116</td>\n",
       "      <td>99</td>\n",
       "      <td>154</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>46</td>\n",
       "      <td>53</td>\n",
       "      <td>61</td>\n",
       "      <td>29</td>\n",
       "      <td>101</td>\n",
       "      <td>130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>46</td>\n",
       "      <td>86</td>\n",
       "      <td>153</td>\n",
       "      <td>91</td>\n",
       "      <td>65</td>\n",
       "      <td>111</td>\n",
       "      <td>133</td>\n",
       "      <td>80</td>\n",
       "      <td>149</td>\n",
       "      <td>158</td>\n",
       "      <td>143</td>\n",
       "      <td>139</td>\n",
       "      <td>68</td>\n",
       "      <td>46</td>\n",
       "      <td>0</td>\n",
       "      <td>94</td>\n",
       "      <td>23</td>\n",
       "      <td>42</td>\n",
       "      <td>142</td>\n",
       "      <td>174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <td>121</td>\n",
       "      <td>26</td>\n",
       "      <td>107</td>\n",
       "      <td>103</td>\n",
       "      <td>78</td>\n",
       "      <td>106</td>\n",
       "      <td>97</td>\n",
       "      <td>127</td>\n",
       "      <td>102</td>\n",
       "      <td>102</td>\n",
       "      <td>85</td>\n",
       "      <td>202</td>\n",
       "      <td>32</td>\n",
       "      <td>53</td>\n",
       "      <td>94</td>\n",
       "      <td>0</td>\n",
       "      <td>101</td>\n",
       "      <td>56</td>\n",
       "      <td>95</td>\n",
       "      <td>110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>65</td>\n",
       "      <td>96</td>\n",
       "      <td>172</td>\n",
       "      <td>108</td>\n",
       "      <td>85</td>\n",
       "      <td>126</td>\n",
       "      <td>152</td>\n",
       "      <td>94</td>\n",
       "      <td>168</td>\n",
       "      <td>177</td>\n",
       "      <td>160</td>\n",
       "      <td>154</td>\n",
       "      <td>81</td>\n",
       "      <td>61</td>\n",
       "      <td>23</td>\n",
       "      <td>101</td>\n",
       "      <td>0</td>\n",
       "      <td>45</td>\n",
       "      <td>160</td>\n",
       "      <td>181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>80</td>\n",
       "      <td>54</td>\n",
       "      <td>138</td>\n",
       "      <td>93</td>\n",
       "      <td>65</td>\n",
       "      <td>107</td>\n",
       "      <td>121</td>\n",
       "      <td>98</td>\n",
       "      <td>134</td>\n",
       "      <td>140</td>\n",
       "      <td>122</td>\n",
       "      <td>170</td>\n",
       "      <td>40</td>\n",
       "      <td>29</td>\n",
       "      <td>42</td>\n",
       "      <td>56</td>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>126</td>\n",
       "      <td>152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>135</td>\n",
       "      <td>91</td>\n",
       "      <td>15</td>\n",
       "      <td>87</td>\n",
       "      <td>81</td>\n",
       "      <td>84</td>\n",
       "      <td>21</td>\n",
       "      <td>127</td>\n",
       "      <td>10</td>\n",
       "      <td>28</td>\n",
       "      <td>36</td>\n",
       "      <td>196</td>\n",
       "      <td>87</td>\n",
       "      <td>101</td>\n",
       "      <td>142</td>\n",
       "      <td>95</td>\n",
       "      <td>160</td>\n",
       "      <td>126</td>\n",
       "      <td>0</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W</th>\n",
       "      <td>177</td>\n",
       "      <td>101</td>\n",
       "      <td>61</td>\n",
       "      <td>147</td>\n",
       "      <td>128</td>\n",
       "      <td>148</td>\n",
       "      <td>88</td>\n",
       "      <td>184</td>\n",
       "      <td>61</td>\n",
       "      <td>40</td>\n",
       "      <td>37</td>\n",
       "      <td>215</td>\n",
       "      <td>115</td>\n",
       "      <td>130</td>\n",
       "      <td>174</td>\n",
       "      <td>110</td>\n",
       "      <td>181</td>\n",
       "      <td>152</td>\n",
       "      <td>67</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     S    R    L    P    T    A    V    G    I    F    Y    C    H    Q    N  \\\n",
       "S    0  110  145   74   58   99  124   56  142  155  144  112   89   68   46   \n",
       "R  110    0  102  103   71  112   96  125   97   97   77  180   29   43   86   \n",
       "L  145  102    0   98   92   96   32  138    5   22   36  198   99  113  153   \n",
       "P   74  103   98    0   38   27   68   42   95  114  110  169   77   76   91   \n",
       "T   58   71   92   38    0   58   69   59   89  103   92  149   47   42   65   \n",
       "A   99  112   96   27   58    0   64   60   94  113  112  195   86   91  111   \n",
       "V  124   96   32   68   69   64    0  109   29   50   55  192   84   96  133   \n",
       "G   56  125  138   42   59   60  109    0  135  153  147  159   98   87   80   \n",
       "I  142   97    5   95   89   94   29  135    0   21   33  198   94  109  149   \n",
       "F  155   97   22  114  103  113   50  153   21    0   22  205  100  116  158   \n",
       "Y  144   77   36  110   92  112   55  147   33   22    0  194   83   99  143   \n",
       "C  112  180  198  169  149  195  192  159  198  205  194    0  174  154  139   \n",
       "H   89   29   99   77   47   86   84   98   94  100   83  174    0   24   68   \n",
       "Q   68   43  113   76   42   91   96   87  109  116   99  154   24    0   46   \n",
       "N   46   86  153   91   65  111  133   80  149  158  143  139   68   46    0   \n",
       "K  121   26  107  103   78  106   97  127  102  102   85  202   32   53   94   \n",
       "D   65   96  172  108   85  126  152   94  168  177  160  154   81   61   23   \n",
       "E   80   54  138   93   65  107  121   98  134  140  122  170   40   29   42   \n",
       "M  135   91   15   87   81   84   21  127   10   28   36  196   87  101  142   \n",
       "W  177  101   61  147  128  148   88  184   61   40   37  215  115  130  174   \n",
       "\n",
       "     K    D    E    M    W  \n",
       "S  121   65   80  135  177  \n",
       "R   26   96   54   91  101  \n",
       "L  107  172  138   15   61  \n",
       "P  103  108   93   87  147  \n",
       "T   78   85   65   81  128  \n",
       "A  106  126  107   84  148  \n",
       "V   97  152  121   21   88  \n",
       "G  127   94   98  127  184  \n",
       "I  102  168  134   10   61  \n",
       "F  102  177  140   28   40  \n",
       "Y   85  160  122   36   37  \n",
       "C  202  154  170  196  215  \n",
       "H   32   81   40   87  115  \n",
       "Q   53   61   29  101  130  \n",
       "N   94   23   42  142  174  \n",
       "K    0  101   56   95  110  \n",
       "D  101    0   45  160  181  \n",
       "E   56   45    0  126  152  \n",
       "M   95  160  126    0   67  \n",
       "W  110  181  152   67    0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# number of nitrogen atoms; Shenhav and Zeevi, Science, 2020\n",
    "nitrogen_atoms = {'A': 1, 'C': 1, 'D': 1, 'E': 1, 'F': 1, 'G': 1, 'H': 3, 'I': 1, 'K': 2, 'L': 1, 'M':1, 'N': 2, \n",
    "    'P': 1, 'Q': 2, 'R': 4, 'S': 1, 'T': 1, 'V': 1, 'W': 2, 'Y': 1}\n",
    "# number of carbon atoms; Shenhav and Zeevi, Science, 2020\n",
    "carbon_atoms = {'A': 3, 'C': 3, 'D': 4, 'E': 5, 'F': 9, 'G': 2, 'H': 6, 'I': 6, 'K': 6, 'L': 6, 'M':5, 'N': 4, \n",
    "    'P': 5, 'Q': 5, 'R': 6, 'S': 3, 'T': 4, 'V': 5, 'W': 11, 'Y': 9}\n",
    "# polar requirement; Woese, Cold Spring Harbor Symposium on Quantitative Biology, 1966\n",
    "polar_requirement = {'A': 7.0, 'C': 4.8, 'D': 13.0, 'E': 12.5, 'F': 5.0, 'G': 7.9, 'H': 8.4, 'I': 4.9, \n",
    "    'K': 10.1, 'L': 4.9, 'M': 5.3, 'N': 10.0, 'P': 6.6, 'Q': 8.6, 'R': 9.1, 'S': 7.5, 'T': 6.6, 'V': 5.6, \n",
    "    'W': 5.2, 'Y': 5.4}\n",
    "# hydropathy; Kyte and Doolittle, J Mol Biol, 1982\n",
    "hydropathy = {'A': 1.8, 'C': 2.5, 'D': -3.5, 'E': -3.5, 'F': 2.8, 'G': -0.4, 'H': -3.2, 'I': 4.5, \n",
    "    'K': -3.9, 'L': 3.8, 'M': 1.9, 'N': -3.5, 'P': -1.6, 'Q': -3.5, 'R': -4.5, 'S': -0.8, 'T': -0.7, 'V': 4.2, \n",
    "    'W': -0.9, 'Y': -1.3}\n",
    "# molecular volume; Grantham, Science, 1974\n",
    "mol_volume = {'A': 31, 'C': 55, 'D': 54, 'E': 83, 'F': 132, 'G': 3, 'H': 96, 'I': 111, \n",
    "    'K': 119, 'L': 111, 'M': 105, 'N': 56, 'P': 32.5, 'Q': 85, 'R': 124, 'S': 32, 'T': 61, 'V': 84, \n",
    "    'W': 170, 'Y': 136}\n",
    "# the Grantham matrix of amino acid distances; Grantham, Science, 1974\n",
    "grantham_matrix = pandas.read_csv(\"../data/grantham.tsv\", sep=\"\\t\", index_col = 0)\n",
    "grantham_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "primary-mitchell",
   "metadata": {},
   "source": [
    "### Codon abundances from diverse species"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "conventional-temple",
   "metadata": {},
   "source": [
    "Codon abundances from 39 organisms, as used by Shenhav and Zeevi (original source of the data: Athey, BMC Bioinformatics, 2017)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "moderate-final",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTT</th>\n",
       "      <th>TTC</th>\n",
       "      <th>TTA</th>\n",
       "      <th>TTG</th>\n",
       "      <th>CTT</th>\n",
       "      <th>CTC</th>\n",
       "      <th>CTA</th>\n",
       "      <th>CTG</th>\n",
       "      <th>ATT</th>\n",
       "      <th>ATC</th>\n",
       "      <th>...</th>\n",
       "      <th>CGA</th>\n",
       "      <th>CGG</th>\n",
       "      <th>AGT</th>\n",
       "      <th>AGC</th>\n",
       "      <th>AGA</th>\n",
       "      <th>AGG</th>\n",
       "      <th>GGT</th>\n",
       "      <th>GGC</th>\n",
       "      <th>GGA</th>\n",
       "      <th>GGG</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Species</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Escherichia coli</th>\n",
       "      <td>0.022272</td>\n",
       "      <td>0.016333</td>\n",
       "      <td>0.013623</td>\n",
       "      <td>0.013200</td>\n",
       "      <td>0.011678</td>\n",
       "      <td>0.010955</td>\n",
       "      <td>0.003977</td>\n",
       "      <td>0.051326</td>\n",
       "      <td>0.029764</td>\n",
       "      <td>0.024562</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003796</td>\n",
       "      <td>0.006062</td>\n",
       "      <td>0.009205</td>\n",
       "      <td>0.015794</td>\n",
       "      <td>0.002711</td>\n",
       "      <td>0.001689</td>\n",
       "      <td>0.024344</td>\n",
       "      <td>0.028417</td>\n",
       "      <td>0.008636</td>\n",
       "      <td>0.011247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacillus subtilis</th>\n",
       "      <td>0.029902</td>\n",
       "      <td>0.014492</td>\n",
       "      <td>0.018661</td>\n",
       "      <td>0.014795</td>\n",
       "      <td>0.022979</td>\n",
       "      <td>0.011292</td>\n",
       "      <td>0.004348</td>\n",
       "      <td>0.024125</td>\n",
       "      <td>0.036057</td>\n",
       "      <td>0.027605</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003781</td>\n",
       "      <td>0.007026</td>\n",
       "      <td>0.006406</td>\n",
       "      <td>0.014194</td>\n",
       "      <td>0.010606</td>\n",
       "      <td>0.003813</td>\n",
       "      <td>0.012549</td>\n",
       "      <td>0.024139</td>\n",
       "      <td>0.021719</td>\n",
       "      <td>0.011159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Natronobacterium gregoryi SP2</th>\n",
       "      <td>0.005295</td>\n",
       "      <td>0.026932</td>\n",
       "      <td>0.001506</td>\n",
       "      <td>0.006181</td>\n",
       "      <td>0.005745</td>\n",
       "      <td>0.048599</td>\n",
       "      <td>0.002853</td>\n",
       "      <td>0.024215</td>\n",
       "      <td>0.005318</td>\n",
       "      <td>0.035001</td>\n",
       "      <td>...</td>\n",
       "      <td>0.013543</td>\n",
       "      <td>0.018691</td>\n",
       "      <td>0.005648</td>\n",
       "      <td>0.012255</td>\n",
       "      <td>0.002168</td>\n",
       "      <td>0.001805</td>\n",
       "      <td>0.014192</td>\n",
       "      <td>0.038010</td>\n",
       "      <td>0.012249</td>\n",
       "      <td>0.015762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Streptomyces coelicolor A3(2)</th>\n",
       "      <td>0.000441</td>\n",
       "      <td>0.025880</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>0.002428</td>\n",
       "      <td>0.001609</td>\n",
       "      <td>0.036529</td>\n",
       "      <td>0.000359</td>\n",
       "      <td>0.060785</td>\n",
       "      <td>0.000634</td>\n",
       "      <td>0.027669</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002589</td>\n",
       "      <td>0.031708</td>\n",
       "      <td>0.001512</td>\n",
       "      <td>0.012404</td>\n",
       "      <td>0.000776</td>\n",
       "      <td>0.003648</td>\n",
       "      <td>0.009214</td>\n",
       "      <td>0.061030</td>\n",
       "      <td>0.007141</td>\n",
       "      <td>0.018068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Deinococcus radiodurans R1</th>\n",
       "      <td>0.010590</td>\n",
       "      <td>0.020760</td>\n",
       "      <td>0.000677</td>\n",
       "      <td>0.007422</td>\n",
       "      <td>0.004948</td>\n",
       "      <td>0.034218</td>\n",
       "      <td>0.001372</td>\n",
       "      <td>0.068345</td>\n",
       "      <td>0.010019</td>\n",
       "      <td>0.022081</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002197</td>\n",
       "      <td>0.020535</td>\n",
       "      <td>0.004529</td>\n",
       "      <td>0.023968</td>\n",
       "      <td>0.000965</td>\n",
       "      <td>0.001995</td>\n",
       "      <td>0.007219</td>\n",
       "      <td>0.058150</td>\n",
       "      <td>0.005437</td>\n",
       "      <td>0.020614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Azotobacter vinelandii</th>\n",
       "      <td>0.002740</td>\n",
       "      <td>0.032100</td>\n",
       "      <td>0.000451</td>\n",
       "      <td>0.006903</td>\n",
       "      <td>0.003082</td>\n",
       "      <td>0.024052</td>\n",
       "      <td>0.001298</td>\n",
       "      <td>0.069679</td>\n",
       "      <td>0.004335</td>\n",
       "      <td>0.042346</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002172</td>\n",
       "      <td>0.010859</td>\n",
       "      <td>0.003190</td>\n",
       "      <td>0.024152</td>\n",
       "      <td>0.000829</td>\n",
       "      <td>0.002424</td>\n",
       "      <td>0.012301</td>\n",
       "      <td>0.069409</td>\n",
       "      <td>0.004380</td>\n",
       "      <td>0.008922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mycoplasma genitalium</th>\n",
       "      <td>0.036439</td>\n",
       "      <td>0.011406</td>\n",
       "      <td>0.031889</td>\n",
       "      <td>0.014841</td>\n",
       "      <td>0.013445</td>\n",
       "      <td>0.008107</td>\n",
       "      <td>0.013409</td>\n",
       "      <td>0.005048</td>\n",
       "      <td>0.027199</td>\n",
       "      <td>0.022173</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001128</td>\n",
       "      <td>0.001119</td>\n",
       "      <td>0.047514</td>\n",
       "      <td>0.008669</td>\n",
       "      <td>0.012770</td>\n",
       "      <td>0.006675</td>\n",
       "      <td>0.021639</td>\n",
       "      <td>0.008746</td>\n",
       "      <td>0.014465</td>\n",
       "      <td>0.018176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Acidianus brierleyi</th>\n",
       "      <td>0.033612</td>\n",
       "      <td>0.013404</td>\n",
       "      <td>0.045393</td>\n",
       "      <td>0.010862</td>\n",
       "      <td>0.018228</td>\n",
       "      <td>0.004011</td>\n",
       "      <td>0.016753</td>\n",
       "      <td>0.003162</td>\n",
       "      <td>0.035097</td>\n",
       "      <td>0.006599</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001058</td>\n",
       "      <td>0.000245</td>\n",
       "      <td>0.014623</td>\n",
       "      <td>0.006156</td>\n",
       "      <td>0.026651</td>\n",
       "      <td>0.008375</td>\n",
       "      <td>0.018843</td>\n",
       "      <td>0.006470</td>\n",
       "      <td>0.033579</td>\n",
       "      <td>0.005159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pyrolobus fumarii 1A</th>\n",
       "      <td>0.006663</td>\n",
       "      <td>0.019546</td>\n",
       "      <td>0.005175</td>\n",
       "      <td>0.012536</td>\n",
       "      <td>0.017240</td>\n",
       "      <td>0.031382</td>\n",
       "      <td>0.026743</td>\n",
       "      <td>0.020520</td>\n",
       "      <td>0.009247</td>\n",
       "      <td>0.016971</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003686</td>\n",
       "      <td>0.005500</td>\n",
       "      <td>0.008733</td>\n",
       "      <td>0.018347</td>\n",
       "      <td>0.013736</td>\n",
       "      <td>0.029626</td>\n",
       "      <td>0.021228</td>\n",
       "      <td>0.031076</td>\n",
       "      <td>0.008383</td>\n",
       "      <td>0.014051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pyrococcus abyssi GE5</th>\n",
       "      <td>0.016319</td>\n",
       "      <td>0.026970</td>\n",
       "      <td>0.016677</td>\n",
       "      <td>0.014596</td>\n",
       "      <td>0.021314</td>\n",
       "      <td>0.020683</td>\n",
       "      <td>0.018430</td>\n",
       "      <td>0.010568</td>\n",
       "      <td>0.021121</td>\n",
       "      <td>0.015419</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000878</td>\n",
       "      <td>0.000629</td>\n",
       "      <td>0.007451</td>\n",
       "      <td>0.013941</td>\n",
       "      <td>0.016805</td>\n",
       "      <td>0.036557</td>\n",
       "      <td>0.015573</td>\n",
       "      <td>0.010826</td>\n",
       "      <td>0.030015</td>\n",
       "      <td>0.015971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Synechococcus lividus PCC 6715</th>\n",
       "      <td>0.027032</td>\n",
       "      <td>0.007750</td>\n",
       "      <td>0.015702</td>\n",
       "      <td>0.021664</td>\n",
       "      <td>0.014029</td>\n",
       "      <td>0.022179</td>\n",
       "      <td>0.014609</td>\n",
       "      <td>0.031371</td>\n",
       "      <td>0.037623</td>\n",
       "      <td>0.015746</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006328</td>\n",
       "      <td>0.015293</td>\n",
       "      <td>0.012979</td>\n",
       "      <td>0.013731</td>\n",
       "      <td>0.001192</td>\n",
       "      <td>0.001668</td>\n",
       "      <td>0.015278</td>\n",
       "      <td>0.025661</td>\n",
       "      <td>0.008130</td>\n",
       "      <td>0.021440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drosophila melanogaster</th>\n",
       "      <td>0.013140</td>\n",
       "      <td>0.021513</td>\n",
       "      <td>0.004450</td>\n",
       "      <td>0.016058</td>\n",
       "      <td>0.008765</td>\n",
       "      <td>0.013618</td>\n",
       "      <td>0.008169</td>\n",
       "      <td>0.038167</td>\n",
       "      <td>0.016534</td>\n",
       "      <td>0.022771</td>\n",
       "      <td>...</td>\n",
       "      <td>0.008574</td>\n",
       "      <td>0.008112</td>\n",
       "      <td>0.011583</td>\n",
       "      <td>0.020768</td>\n",
       "      <td>0.005153</td>\n",
       "      <td>0.006256</td>\n",
       "      <td>0.013375</td>\n",
       "      <td>0.026718</td>\n",
       "      <td>0.017880</td>\n",
       "      <td>0.004639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Caenorhabditis elegans</th>\n",
       "      <td>0.018926</td>\n",
       "      <td>0.022956</td>\n",
       "      <td>0.008319</td>\n",
       "      <td>0.019441</td>\n",
       "      <td>0.021692</td>\n",
       "      <td>0.015936</td>\n",
       "      <td>0.006944</td>\n",
       "      <td>0.012138</td>\n",
       "      <td>0.030410</td>\n",
       "      <td>0.019499</td>\n",
       "      <td>...</td>\n",
       "      <td>0.012958</td>\n",
       "      <td>0.004743</td>\n",
       "      <td>0.011689</td>\n",
       "      <td>0.007957</td>\n",
       "      <td>0.015013</td>\n",
       "      <td>0.003636</td>\n",
       "      <td>0.011753</td>\n",
       "      <td>0.006468</td>\n",
       "      <td>0.035210</td>\n",
       "      <td>0.003825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tribolium castaneum</th>\n",
       "      <td>0.024160</td>\n",
       "      <td>0.026753</td>\n",
       "      <td>0.013356</td>\n",
       "      <td>0.026075</td>\n",
       "      <td>0.012313</td>\n",
       "      <td>0.017473</td>\n",
       "      <td>0.008883</td>\n",
       "      <td>0.018097</td>\n",
       "      <td>0.026406</td>\n",
       "      <td>0.024673</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007608</td>\n",
       "      <td>0.007597</td>\n",
       "      <td>0.012674</td>\n",
       "      <td>0.013116</td>\n",
       "      <td>0.009385</td>\n",
       "      <td>0.008974</td>\n",
       "      <td>0.011648</td>\n",
       "      <td>0.017081</td>\n",
       "      <td>0.013171</td>\n",
       "      <td>0.013516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Felis catus</th>\n",
       "      <td>0.014825</td>\n",
       "      <td>0.025656</td>\n",
       "      <td>0.005915</td>\n",
       "      <td>0.011083</td>\n",
       "      <td>0.010418</td>\n",
       "      <td>0.022609</td>\n",
       "      <td>0.005973</td>\n",
       "      <td>0.043244</td>\n",
       "      <td>0.013154</td>\n",
       "      <td>0.023524</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005128</td>\n",
       "      <td>0.011205</td>\n",
       "      <td>0.011245</td>\n",
       "      <td>0.021884</td>\n",
       "      <td>0.011055</td>\n",
       "      <td>0.011886</td>\n",
       "      <td>0.009802</td>\n",
       "      <td>0.023086</td>\n",
       "      <td>0.016272</td>\n",
       "      <td>0.018202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Canis lupus familiaris</th>\n",
       "      <td>0.017515</td>\n",
       "      <td>0.026588</td>\n",
       "      <td>0.006387</td>\n",
       "      <td>0.012489</td>\n",
       "      <td>0.011499</td>\n",
       "      <td>0.023746</td>\n",
       "      <td>0.006233</td>\n",
       "      <td>0.043175</td>\n",
       "      <td>0.016797</td>\n",
       "      <td>0.025633</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005683</td>\n",
       "      <td>0.012012</td>\n",
       "      <td>0.010455</td>\n",
       "      <td>0.019748</td>\n",
       "      <td>0.010997</td>\n",
       "      <td>0.010891</td>\n",
       "      <td>0.010706</td>\n",
       "      <td>0.023305</td>\n",
       "      <td>0.014515</td>\n",
       "      <td>0.017677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Arabidopsis thaliana</th>\n",
       "      <td>0.021992</td>\n",
       "      <td>0.020106</td>\n",
       "      <td>0.013189</td>\n",
       "      <td>0.021400</td>\n",
       "      <td>0.024649</td>\n",
       "      <td>0.015979</td>\n",
       "      <td>0.010288</td>\n",
       "      <td>0.010275</td>\n",
       "      <td>0.021595</td>\n",
       "      <td>0.018132</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006277</td>\n",
       "      <td>0.004872</td>\n",
       "      <td>0.014529</td>\n",
       "      <td>0.011434</td>\n",
       "      <td>0.019238</td>\n",
       "      <td>0.011045</td>\n",
       "      <td>0.021563</td>\n",
       "      <td>0.008968</td>\n",
       "      <td>0.023635</td>\n",
       "      <td>0.010142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Aspergillus nidulans</th>\n",
       "      <td>0.013350</td>\n",
       "      <td>0.025755</td>\n",
       "      <td>0.004726</td>\n",
       "      <td>0.013102</td>\n",
       "      <td>0.019086</td>\n",
       "      <td>0.023561</td>\n",
       "      <td>0.007962</td>\n",
       "      <td>0.019696</td>\n",
       "      <td>0.018852</td>\n",
       "      <td>0.026902</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009128</td>\n",
       "      <td>0.009475</td>\n",
       "      <td>0.008882</td>\n",
       "      <td>0.015456</td>\n",
       "      <td>0.005767</td>\n",
       "      <td>0.005906</td>\n",
       "      <td>0.020633</td>\n",
       "      <td>0.025152</td>\n",
       "      <td>0.014415</td>\n",
       "      <td>0.010541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Saccharomyces cerevisiae</th>\n",
       "      <td>0.025150</td>\n",
       "      <td>0.018566</td>\n",
       "      <td>0.025311</td>\n",
       "      <td>0.027447</td>\n",
       "      <td>0.011944</td>\n",
       "      <td>0.005192</td>\n",
       "      <td>0.013003</td>\n",
       "      <td>0.010251</td>\n",
       "      <td>0.029953</td>\n",
       "      <td>0.017317</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002988</td>\n",
       "      <td>0.001711</td>\n",
       "      <td>0.014194</td>\n",
       "      <td>0.009447</td>\n",
       "      <td>0.021161</td>\n",
       "      <td>0.008857</td>\n",
       "      <td>0.025169</td>\n",
       "      <td>0.010006</td>\n",
       "      <td>0.010591</td>\n",
       "      <td>0.006111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Schizosaccharomyces pombe</th>\n",
       "      <td>0.032459</td>\n",
       "      <td>0.013107</td>\n",
       "      <td>0.026138</td>\n",
       "      <td>0.024115</td>\n",
       "      <td>0.025238</td>\n",
       "      <td>0.007323</td>\n",
       "      <td>0.008722</td>\n",
       "      <td>0.006448</td>\n",
       "      <td>0.034989</td>\n",
       "      <td>0.012668</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007990</td>\n",
       "      <td>0.002985</td>\n",
       "      <td>0.014879</td>\n",
       "      <td>0.009229</td>\n",
       "      <td>0.011284</td>\n",
       "      <td>0.005107</td>\n",
       "      <td>0.021596</td>\n",
       "      <td>0.008316</td>\n",
       "      <td>0.015847</td>\n",
       "      <td>0.004383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zea mays</th>\n",
       "      <td>0.011103</td>\n",
       "      <td>0.024943</td>\n",
       "      <td>0.004336</td>\n",
       "      <td>0.011666</td>\n",
       "      <td>0.014045</td>\n",
       "      <td>0.027614</td>\n",
       "      <td>0.006233</td>\n",
       "      <td>0.027359</td>\n",
       "      <td>0.011557</td>\n",
       "      <td>0.020938</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004584</td>\n",
       "      <td>0.012606</td>\n",
       "      <td>0.007304</td>\n",
       "      <td>0.017952</td>\n",
       "      <td>0.008403</td>\n",
       "      <td>0.015911</td>\n",
       "      <td>0.013636</td>\n",
       "      <td>0.032336</td>\n",
       "      <td>0.013510</td>\n",
       "      <td>0.017021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Chlamydomonas reinhardtii</th>\n",
       "      <td>0.005120</td>\n",
       "      <td>0.024609</td>\n",
       "      <td>0.000748</td>\n",
       "      <td>0.004580</td>\n",
       "      <td>0.004489</td>\n",
       "      <td>0.013021</td>\n",
       "      <td>0.002994</td>\n",
       "      <td>0.065145</td>\n",
       "      <td>0.007270</td>\n",
       "      <td>0.024309</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002286</td>\n",
       "      <td>0.013241</td>\n",
       "      <td>0.003203</td>\n",
       "      <td>0.024109</td>\n",
       "      <td>0.000833</td>\n",
       "      <td>0.003232</td>\n",
       "      <td>0.010203</td>\n",
       "      <td>0.063039</td>\n",
       "      <td>0.005587</td>\n",
       "      <td>0.011586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Oryza sativa</th>\n",
       "      <td>0.013923</td>\n",
       "      <td>0.021637</td>\n",
       "      <td>0.007180</td>\n",
       "      <td>0.015520</td>\n",
       "      <td>0.017623</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>0.008961</td>\n",
       "      <td>0.023435</td>\n",
       "      <td>0.016526</td>\n",
       "      <td>0.020779</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005373</td>\n",
       "      <td>0.010594</td>\n",
       "      <td>0.009272</td>\n",
       "      <td>0.016837</td>\n",
       "      <td>0.010685</td>\n",
       "      <td>0.014601</td>\n",
       "      <td>0.014947</td>\n",
       "      <td>0.026761</td>\n",
       "      <td>0.014789</td>\n",
       "      <td>0.015607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lotus japonicus</th>\n",
       "      <td>0.023690</td>\n",
       "      <td>0.020546</td>\n",
       "      <td>0.010878</td>\n",
       "      <td>0.023015</td>\n",
       "      <td>0.023816</td>\n",
       "      <td>0.015150</td>\n",
       "      <td>0.008971</td>\n",
       "      <td>0.013110</td>\n",
       "      <td>0.025137</td>\n",
       "      <td>0.015685</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004807</td>\n",
       "      <td>0.003694</td>\n",
       "      <td>0.013349</td>\n",
       "      <td>0.011818</td>\n",
       "      <td>0.015887</td>\n",
       "      <td>0.013951</td>\n",
       "      <td>0.020569</td>\n",
       "      <td>0.011238</td>\n",
       "      <td>0.021229</td>\n",
       "      <td>0.012569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dunaliella salina</th>\n",
       "      <td>0.011706</td>\n",
       "      <td>0.025895</td>\n",
       "      <td>0.002054</td>\n",
       "      <td>0.012321</td>\n",
       "      <td>0.006764</td>\n",
       "      <td>0.013342</td>\n",
       "      <td>0.003515</td>\n",
       "      <td>0.048948</td>\n",
       "      <td>0.013528</td>\n",
       "      <td>0.030153</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004536</td>\n",
       "      <td>0.006393</td>\n",
       "      <td>0.004490</td>\n",
       "      <td>0.019305</td>\n",
       "      <td>0.003028</td>\n",
       "      <td>0.010279</td>\n",
       "      <td>0.015233</td>\n",
       "      <td>0.041546</td>\n",
       "      <td>0.009803</td>\n",
       "      <td>0.010685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Physcomitrella patens</th>\n",
       "      <td>0.017719</td>\n",
       "      <td>0.020673</td>\n",
       "      <td>0.008779</td>\n",
       "      <td>0.023700</td>\n",
       "      <td>0.018065</td>\n",
       "      <td>0.013240</td>\n",
       "      <td>0.008081</td>\n",
       "      <td>0.021467</td>\n",
       "      <td>0.020051</td>\n",
       "      <td>0.018465</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009926</td>\n",
       "      <td>0.009905</td>\n",
       "      <td>0.012490</td>\n",
       "      <td>0.015011</td>\n",
       "      <td>0.010792</td>\n",
       "      <td>0.011658</td>\n",
       "      <td>0.018452</td>\n",
       "      <td>0.017262</td>\n",
       "      <td>0.022077</td>\n",
       "      <td>0.015831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Emiliania huxleyi</th>\n",
       "      <td>0.010187</td>\n",
       "      <td>0.027770</td>\n",
       "      <td>0.003858</td>\n",
       "      <td>0.003556</td>\n",
       "      <td>0.004516</td>\n",
       "      <td>0.049994</td>\n",
       "      <td>0.003040</td>\n",
       "      <td>0.031664</td>\n",
       "      <td>0.010098</td>\n",
       "      <td>0.034757</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005085</td>\n",
       "      <td>0.013530</td>\n",
       "      <td>0.001760</td>\n",
       "      <td>0.010969</td>\n",
       "      <td>0.001013</td>\n",
       "      <td>0.002240</td>\n",
       "      <td>0.011556</td>\n",
       "      <td>0.049620</td>\n",
       "      <td>0.007716</td>\n",
       "      <td>0.012054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Thalassiosira pseudonana</th>\n",
       "      <td>0.014819</td>\n",
       "      <td>0.022527</td>\n",
       "      <td>0.014915</td>\n",
       "      <td>0.019440</td>\n",
       "      <td>0.014463</td>\n",
       "      <td>0.019182</td>\n",
       "      <td>0.007466</td>\n",
       "      <td>0.007934</td>\n",
       "      <td>0.021266</td>\n",
       "      <td>0.025322</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009130</td>\n",
       "      <td>0.003297</td>\n",
       "      <td>0.011667</td>\n",
       "      <td>0.008209</td>\n",
       "      <td>0.008597</td>\n",
       "      <td>0.006690</td>\n",
       "      <td>0.022511</td>\n",
       "      <td>0.012007</td>\n",
       "      <td>0.025322</td>\n",
       "      <td>0.007434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Populus trichocarpa</th>\n",
       "      <td>0.023691</td>\n",
       "      <td>0.019565</td>\n",
       "      <td>0.009902</td>\n",
       "      <td>0.022257</td>\n",
       "      <td>0.024078</td>\n",
       "      <td>0.013735</td>\n",
       "      <td>0.009078</td>\n",
       "      <td>0.012228</td>\n",
       "      <td>0.024310</td>\n",
       "      <td>0.016424</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004989</td>\n",
       "      <td>0.003782</td>\n",
       "      <td>0.012804</td>\n",
       "      <td>0.011643</td>\n",
       "      <td>0.015738</td>\n",
       "      <td>0.014041</td>\n",
       "      <td>0.021371</td>\n",
       "      <td>0.012546</td>\n",
       "      <td>0.021989</td>\n",
       "      <td>0.012965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homo sapiens</th>\n",
       "      <td>0.015155</td>\n",
       "      <td>0.021402</td>\n",
       "      <td>0.006401</td>\n",
       "      <td>0.011481</td>\n",
       "      <td>0.011303</td>\n",
       "      <td>0.019416</td>\n",
       "      <td>0.006053</td>\n",
       "      <td>0.040141</td>\n",
       "      <td>0.013826</td>\n",
       "      <td>0.021328</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006269</td>\n",
       "      <td>0.012040</td>\n",
       "      <td>0.012746</td>\n",
       "      <td>0.020351</td>\n",
       "      <td>0.012992</td>\n",
       "      <td>0.011741</td>\n",
       "      <td>0.010503</td>\n",
       "      <td>0.023176</td>\n",
       "      <td>0.016677</td>\n",
       "      <td>0.020049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pan troglodytes</th>\n",
       "      <td>0.017967</td>\n",
       "      <td>0.021218</td>\n",
       "      <td>0.007842</td>\n",
       "      <td>0.013027</td>\n",
       "      <td>0.013240</td>\n",
       "      <td>0.020238</td>\n",
       "      <td>0.007541</td>\n",
       "      <td>0.038064</td>\n",
       "      <td>0.016762</td>\n",
       "      <td>0.021561</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005993</td>\n",
       "      <td>0.010966</td>\n",
       "      <td>0.012166</td>\n",
       "      <td>0.018944</td>\n",
       "      <td>0.013374</td>\n",
       "      <td>0.012893</td>\n",
       "      <td>0.010009</td>\n",
       "      <td>0.020086</td>\n",
       "      <td>0.015998</td>\n",
       "      <td>0.016705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mus musculus</th>\n",
       "      <td>0.015910</td>\n",
       "      <td>0.021827</td>\n",
       "      <td>0.006424</td>\n",
       "      <td>0.012610</td>\n",
       "      <td>0.012832</td>\n",
       "      <td>0.019703</td>\n",
       "      <td>0.007448</td>\n",
       "      <td>0.038813</td>\n",
       "      <td>0.015197</td>\n",
       "      <td>0.021084</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006380</td>\n",
       "      <td>0.009551</td>\n",
       "      <td>0.013695</td>\n",
       "      <td>0.021015</td>\n",
       "      <td>0.011820</td>\n",
       "      <td>0.012234</td>\n",
       "      <td>0.012358</td>\n",
       "      <td>0.021618</td>\n",
       "      <td>0.017658</td>\n",
       "      <td>0.015905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rattus norvegicus</th>\n",
       "      <td>0.016284</td>\n",
       "      <td>0.022887</td>\n",
       "      <td>0.005937</td>\n",
       "      <td>0.012664</td>\n",
       "      <td>0.012446</td>\n",
       "      <td>0.020071</td>\n",
       "      <td>0.007583</td>\n",
       "      <td>0.040675</td>\n",
       "      <td>0.015121</td>\n",
       "      <td>0.023898</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006826</td>\n",
       "      <td>0.010883</td>\n",
       "      <td>0.012000</td>\n",
       "      <td>0.019189</td>\n",
       "      <td>0.011303</td>\n",
       "      <td>0.011862</td>\n",
       "      <td>0.011491</td>\n",
       "      <td>0.021718</td>\n",
       "      <td>0.016793</td>\n",
       "      <td>0.015602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Heterocephalus glaber</th>\n",
       "      <td>0.017654</td>\n",
       "      <td>0.021853</td>\n",
       "      <td>0.006547</td>\n",
       "      <td>0.012146</td>\n",
       "      <td>0.011570</td>\n",
       "      <td>0.018184</td>\n",
       "      <td>0.005802</td>\n",
       "      <td>0.038976</td>\n",
       "      <td>0.017112</td>\n",
       "      <td>0.023986</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006061</td>\n",
       "      <td>0.010486</td>\n",
       "      <td>0.014855</td>\n",
       "      <td>0.021198</td>\n",
       "      <td>0.010091</td>\n",
       "      <td>0.009956</td>\n",
       "      <td>0.012541</td>\n",
       "      <td>0.022914</td>\n",
       "      <td>0.017394</td>\n",
       "      <td>0.015442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Danio rerio</th>\n",
       "      <td>0.017842</td>\n",
       "      <td>0.020642</td>\n",
       "      <td>0.006819</td>\n",
       "      <td>0.012068</td>\n",
       "      <td>0.012619</td>\n",
       "      <td>0.016856</td>\n",
       "      <td>0.006148</td>\n",
       "      <td>0.037655</td>\n",
       "      <td>0.016365</td>\n",
       "      <td>0.023717</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006604</td>\n",
       "      <td>0.006546</td>\n",
       "      <td>0.013429</td>\n",
       "      <td>0.018634</td>\n",
       "      <td>0.014449</td>\n",
       "      <td>0.010064</td>\n",
       "      <td>0.013902</td>\n",
       "      <td>0.017138</td>\n",
       "      <td>0.022041</td>\n",
       "      <td>0.010038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gallus gallus</th>\n",
       "      <td>0.018623</td>\n",
       "      <td>0.020585</td>\n",
       "      <td>0.007935</td>\n",
       "      <td>0.014071</td>\n",
       "      <td>0.013730</td>\n",
       "      <td>0.017507</td>\n",
       "      <td>0.006909</td>\n",
       "      <td>0.041364</td>\n",
       "      <td>0.017097</td>\n",
       "      <td>0.022846</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004682</td>\n",
       "      <td>0.009233</td>\n",
       "      <td>0.011755</td>\n",
       "      <td>0.020823</td>\n",
       "      <td>0.012221</td>\n",
       "      <td>0.011894</td>\n",
       "      <td>0.010778</td>\n",
       "      <td>0.018616</td>\n",
       "      <td>0.016405</td>\n",
       "      <td>0.016262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Taeniopygia guttata</th>\n",
       "      <td>0.012006</td>\n",
       "      <td>0.039156</td>\n",
       "      <td>0.002973</td>\n",
       "      <td>0.007424</td>\n",
       "      <td>0.006854</td>\n",
       "      <td>0.034590</td>\n",
       "      <td>0.002388</td>\n",
       "      <td>0.063481</td>\n",
       "      <td>0.014430</td>\n",
       "      <td>0.039689</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002147</td>\n",
       "      <td>0.012147</td>\n",
       "      <td>0.006985</td>\n",
       "      <td>0.037443</td>\n",
       "      <td>0.006003</td>\n",
       "      <td>0.011755</td>\n",
       "      <td>0.007131</td>\n",
       "      <td>0.021200</td>\n",
       "      <td>0.014811</td>\n",
       "      <td>0.020312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Xenopus tropicalis</th>\n",
       "      <td>0.021492</td>\n",
       "      <td>0.016609</td>\n",
       "      <td>0.010425</td>\n",
       "      <td>0.014805</td>\n",
       "      <td>0.017409</td>\n",
       "      <td>0.013003</td>\n",
       "      <td>0.009629</td>\n",
       "      <td>0.029077</td>\n",
       "      <td>0.020703</td>\n",
       "      <td>0.017008</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006099</td>\n",
       "      <td>0.007045</td>\n",
       "      <td>0.014122</td>\n",
       "      <td>0.016669</td>\n",
       "      <td>0.014585</td>\n",
       "      <td>0.011962</td>\n",
       "      <td>0.011857</td>\n",
       "      <td>0.014430</td>\n",
       "      <td>0.020429</td>\n",
       "      <td>0.014061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ambystoma mexicanum</th>\n",
       "      <td>0.014331</td>\n",
       "      <td>0.021794</td>\n",
       "      <td>0.005540</td>\n",
       "      <td>0.012893</td>\n",
       "      <td>0.011221</td>\n",
       "      <td>0.017973</td>\n",
       "      <td>0.006400</td>\n",
       "      <td>0.033093</td>\n",
       "      <td>0.016316</td>\n",
       "      <td>0.022395</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007978</td>\n",
       "      <td>0.010924</td>\n",
       "      <td>0.009940</td>\n",
       "      <td>0.020192</td>\n",
       "      <td>0.011104</td>\n",
       "      <td>0.013472</td>\n",
       "      <td>0.011502</td>\n",
       "      <td>0.022036</td>\n",
       "      <td>0.016871</td>\n",
       "      <td>0.018418</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39 rows  64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     TTT       TTC       TTA       TTG  \\\n",
       "Species                                                                  \n",
       "Escherichia coli                0.022272  0.016333  0.013623  0.013200   \n",
       "Bacillus subtilis               0.029902  0.014492  0.018661  0.014795   \n",
       "Natronobacterium gregoryi SP2   0.005295  0.026932  0.001506  0.006181   \n",
       "Streptomyces coelicolor A3(2)   0.000441  0.025880  0.000063  0.002428   \n",
       "Deinococcus radiodurans R1      0.010590  0.020760  0.000677  0.007422   \n",
       "Azotobacter vinelandii          0.002740  0.032100  0.000451  0.006903   \n",
       "Mycoplasma genitalium           0.036439  0.011406  0.031889  0.014841   \n",
       "Acidianus brierleyi             0.033612  0.013404  0.045393  0.010862   \n",
       "Pyrolobus fumarii 1A            0.006663  0.019546  0.005175  0.012536   \n",
       "Pyrococcus abyssi GE5           0.016319  0.026970  0.016677  0.014596   \n",
       "Synechococcus lividus PCC 6715  0.027032  0.007750  0.015702  0.021664   \n",
       "Drosophila melanogaster         0.013140  0.021513  0.004450  0.016058   \n",
       "Caenorhabditis elegans          0.018926  0.022956  0.008319  0.019441   \n",
       "Tribolium castaneum             0.024160  0.026753  0.013356  0.026075   \n",
       "Felis catus                     0.014825  0.025656  0.005915  0.011083   \n",
       "Canis lupus familiaris          0.017515  0.026588  0.006387  0.012489   \n",
       "Arabidopsis thaliana            0.021992  0.020106  0.013189  0.021400   \n",
       "Aspergillus nidulans            0.013350  0.025755  0.004726  0.013102   \n",
       "Saccharomyces cerevisiae        0.025150  0.018566  0.025311  0.027447   \n",
       "Schizosaccharomyces pombe       0.032459  0.013107  0.026138  0.024115   \n",
       "Zea mays                        0.011103  0.024943  0.004336  0.011666   \n",
       "Chlamydomonas reinhardtii       0.005120  0.024609  0.000748  0.004580   \n",
       "Oryza sativa                    0.013923  0.021637  0.007180  0.015520   \n",
       "Lotus japonicus                 0.023690  0.020546  0.010878  0.023015   \n",
       "Dunaliella salina               0.011706  0.025895  0.002054  0.012321   \n",
       "Physcomitrella patens           0.017719  0.020673  0.008779  0.023700   \n",
       "Emiliania huxleyi               0.010187  0.027770  0.003858  0.003556   \n",
       "Thalassiosira pseudonana        0.014819  0.022527  0.014915  0.019440   \n",
       "Populus trichocarpa             0.023691  0.019565  0.009902  0.022257   \n",
       "Homo sapiens                    0.015155  0.021402  0.006401  0.011481   \n",
       "Pan troglodytes                 0.017967  0.021218  0.007842  0.013027   \n",
       "Mus musculus                    0.015910  0.021827  0.006424  0.012610   \n",
       "Rattus norvegicus               0.016284  0.022887  0.005937  0.012664   \n",
       "Heterocephalus glaber           0.017654  0.021853  0.006547  0.012146   \n",
       "Danio rerio                     0.017842  0.020642  0.006819  0.012068   \n",
       "Gallus gallus                   0.018623  0.020585  0.007935  0.014071   \n",
       "Taeniopygia guttata             0.012006  0.039156  0.002973  0.007424   \n",
       "Xenopus tropicalis              0.021492  0.016609  0.010425  0.014805   \n",
       "Ambystoma mexicanum             0.014331  0.021794  0.005540  0.012893   \n",
       "\n",
       "                                     CTT       CTC       CTA       CTG  \\\n",
       "Species                                                                  \n",
       "Escherichia coli                0.011678  0.010955  0.003977  0.051326   \n",
       "Bacillus subtilis               0.022979  0.011292  0.004348  0.024125   \n",
       "Natronobacterium gregoryi SP2   0.005745  0.048599  0.002853  0.024215   \n",
       "Streptomyces coelicolor A3(2)   0.001609  0.036529  0.000359  0.060785   \n",
       "Deinococcus radiodurans R1      0.004948  0.034218  0.001372  0.068345   \n",
       "Azotobacter vinelandii          0.003082  0.024052  0.001298  0.069679   \n",
       "Mycoplasma genitalium           0.013445  0.008107  0.013409  0.005048   \n",
       "Acidianus brierleyi             0.018228  0.004011  0.016753  0.003162   \n",
       "Pyrolobus fumarii 1A            0.017240  0.031382  0.026743  0.020520   \n",
       "Pyrococcus abyssi GE5           0.021314  0.020683  0.018430  0.010568   \n",
       "Synechococcus lividus PCC 6715  0.014029  0.022179  0.014609  0.031371   \n",
       "Drosophila melanogaster         0.008765  0.013618  0.008169  0.038167   \n",
       "Caenorhabditis elegans          0.021692  0.015936  0.006944  0.012138   \n",
       "Tribolium castaneum             0.012313  0.017473  0.008883  0.018097   \n",
       "Felis catus                     0.010418  0.022609  0.005973  0.043244   \n",
       "Canis lupus familiaris          0.011499  0.023746  0.006233  0.043175   \n",
       "Arabidopsis thaliana            0.024649  0.015979  0.010288  0.010275   \n",
       "Aspergillus nidulans            0.019086  0.023561  0.007962  0.019696   \n",
       "Saccharomyces cerevisiae        0.011944  0.005192  0.013003  0.010251   \n",
       "Schizosaccharomyces pombe       0.025238  0.007323  0.008722  0.006448   \n",
       "Zea mays                        0.014045  0.027614  0.006233  0.027359   \n",
       "Chlamydomonas reinhardtii       0.004489  0.013021  0.002994  0.065145   \n",
       "Oryza sativa                    0.017623  0.025641  0.008961  0.023435   \n",
       "Lotus japonicus                 0.023816  0.015150  0.008971  0.013110   \n",
       "Dunaliella salina               0.006764  0.013342  0.003515  0.048948   \n",
       "Physcomitrella patens           0.018065  0.013240  0.008081  0.021467   \n",
       "Emiliania huxleyi               0.004516  0.049994  0.003040  0.031664   \n",
       "Thalassiosira pseudonana        0.014463  0.019182  0.007466  0.007934   \n",
       "Populus trichocarpa             0.024078  0.013735  0.009078  0.012228   \n",
       "Homo sapiens                    0.011303  0.019416  0.006053  0.040141   \n",
       "Pan troglodytes                 0.013240  0.020238  0.007541  0.038064   \n",
       "Mus musculus                    0.012832  0.019703  0.007448  0.038813   \n",
       "Rattus norvegicus               0.012446  0.020071  0.007583  0.040675   \n",
       "Heterocephalus glaber           0.011570  0.018184  0.005802  0.038976   \n",
       "Danio rerio                     0.012619  0.016856  0.006148  0.037655   \n",
       "Gallus gallus                   0.013730  0.017507  0.006909  0.041364   \n",
       "Taeniopygia guttata             0.006854  0.034590  0.002388  0.063481   \n",
       "Xenopus tropicalis              0.017409  0.013003  0.009629  0.029077   \n",
       "Ambystoma mexicanum             0.011221  0.017973  0.006400  0.033093   \n",
       "\n",
       "                                     ATT       ATC  ...       CGA       CGG  \\\n",
       "Species                                             ...                       \n",
       "Escherichia coli                0.029764  0.024562  ...  0.003796  0.006062   \n",
       "Bacillus subtilis               0.036057  0.027605  ...  0.003781  0.007026   \n",
       "Natronobacterium gregoryi SP2   0.005318  0.035001  ...  0.013543  0.018691   \n",
       "Streptomyces coelicolor A3(2)   0.000634  0.027669  ...  0.002589  0.031708   \n",
       "Deinococcus radiodurans R1      0.010019  0.022081  ...  0.002197  0.020535   \n",
       "Azotobacter vinelandii          0.004335  0.042346  ...  0.002172  0.010859   \n",
       "Mycoplasma genitalium           0.027199  0.022173  ...  0.001128  0.001119   \n",
       "Acidianus brierleyi             0.035097  0.006599  ...  0.001058  0.000245   \n",
       "Pyrolobus fumarii 1A            0.009247  0.016971  ...  0.003686  0.005500   \n",
       "Pyrococcus abyssi GE5           0.021121  0.015419  ...  0.000878  0.000629   \n",
       "Synechococcus lividus PCC 6715  0.037623  0.015746  ...  0.006328  0.015293   \n",
       "Drosophila melanogaster         0.016534  0.022771  ...  0.008574  0.008112   \n",
       "Caenorhabditis elegans          0.030410  0.019499  ...  0.012958  0.004743   \n",
       "Tribolium castaneum             0.026406  0.024673  ...  0.007608  0.007597   \n",
       "Felis catus                     0.013154  0.023524  ...  0.005128  0.011205   \n",
       "Canis lupus familiaris          0.016797  0.025633  ...  0.005683  0.012012   \n",
       "Arabidopsis thaliana            0.021595  0.018132  ...  0.006277  0.004872   \n",
       "Aspergillus nidulans            0.018852  0.026902  ...  0.009128  0.009475   \n",
       "Saccharomyces cerevisiae        0.029953  0.017317  ...  0.002988  0.001711   \n",
       "Schizosaccharomyces pombe       0.034989  0.012668  ...  0.007990  0.002985   \n",
       "Zea mays                        0.011557  0.020938  ...  0.004584  0.012606   \n",
       "Chlamydomonas reinhardtii       0.007270  0.024309  ...  0.002286  0.013241   \n",
       "Oryza sativa                    0.016526  0.020779  ...  0.005373  0.010594   \n",
       "Lotus japonicus                 0.025137  0.015685  ...  0.004807  0.003694   \n",
       "Dunaliella salina               0.013528  0.030153  ...  0.004536  0.006393   \n",
       "Physcomitrella patens           0.020051  0.018465  ...  0.009926  0.009905   \n",
       "Emiliania huxleyi               0.010098  0.034757  ...  0.005085  0.013530   \n",
       "Thalassiosira pseudonana        0.021266  0.025322  ...  0.009130  0.003297   \n",
       "Populus trichocarpa             0.024310  0.016424  ...  0.004989  0.003782   \n",
       "Homo sapiens                    0.013826  0.021328  ...  0.006269  0.012040   \n",
       "Pan troglodytes                 0.016762  0.021561  ...  0.005993  0.010966   \n",
       "Mus musculus                    0.015197  0.021084  ...  0.006380  0.009551   \n",
       "Rattus norvegicus               0.015121  0.023898  ...  0.006826  0.010883   \n",
       "Heterocephalus glaber           0.017112  0.023986  ...  0.006061  0.010486   \n",
       "Danio rerio                     0.016365  0.023717  ...  0.006604  0.006546   \n",
       "Gallus gallus                   0.017097  0.022846  ...  0.004682  0.009233   \n",
       "Taeniopygia guttata             0.014430  0.039689  ...  0.002147  0.012147   \n",
       "Xenopus tropicalis              0.020703  0.017008  ...  0.006099  0.007045   \n",
       "Ambystoma mexicanum             0.016316  0.022395  ...  0.007978  0.010924   \n",
       "\n",
       "                                     AGT       AGC       AGA       AGG  \\\n",
       "Species                                                                  \n",
       "Escherichia coli                0.009205  0.015794  0.002711  0.001689   \n",
       "Bacillus subtilis               0.006406  0.014194  0.010606  0.003813   \n",
       "Natronobacterium gregoryi SP2   0.005648  0.012255  0.002168  0.001805   \n",
       "Streptomyces coelicolor A3(2)   0.001512  0.012404  0.000776  0.003648   \n",
       "Deinococcus radiodurans R1      0.004529  0.023968  0.000965  0.001995   \n",
       "Azotobacter vinelandii          0.003190  0.024152  0.000829  0.002424   \n",
       "Mycoplasma genitalium           0.047514  0.008669  0.012770  0.006675   \n",
       "Acidianus brierleyi             0.014623  0.006156  0.026651  0.008375   \n",
       "Pyrolobus fumarii 1A            0.008733  0.018347  0.013736  0.029626   \n",
       "Pyrococcus abyssi GE5           0.007451  0.013941  0.016805  0.036557   \n",
       "Synechococcus lividus PCC 6715  0.012979  0.013731  0.001192  0.001668   \n",
       "Drosophila melanogaster         0.011583  0.020768  0.005153  0.006256   \n",
       "Caenorhabditis elegans          0.011689  0.007957  0.015013  0.003636   \n",
       "Tribolium castaneum             0.012674  0.013116  0.009385  0.008974   \n",
       "Felis catus                     0.011245  0.021884  0.011055  0.011886   \n",
       "Canis lupus familiaris          0.010455  0.019748  0.010997  0.010891   \n",
       "Arabidopsis thaliana            0.014529  0.011434  0.019238  0.011045   \n",
       "Aspergillus nidulans            0.008882  0.015456  0.005767  0.005906   \n",
       "Saccharomyces cerevisiae        0.014194  0.009447  0.021161  0.008857   \n",
       "Schizosaccharomyces pombe       0.014879  0.009229  0.011284  0.005107   \n",
       "Zea mays                        0.007304  0.017952  0.008403  0.015911   \n",
       "Chlamydomonas reinhardtii       0.003203  0.024109  0.000833  0.003232   \n",
       "Oryza sativa                    0.009272  0.016837  0.010685  0.014601   \n",
       "Lotus japonicus                 0.013349  0.011818  0.015887  0.013951   \n",
       "Dunaliella salina               0.004490  0.019305  0.003028  0.010279   \n",
       "Physcomitrella patens           0.012490  0.015011  0.010792  0.011658   \n",
       "Emiliania huxleyi               0.001760  0.010969  0.001013  0.002240   \n",
       "Thalassiosira pseudonana        0.011667  0.008209  0.008597  0.006690   \n",
       "Populus trichocarpa             0.012804  0.011643  0.015738  0.014041   \n",
       "Homo sapiens                    0.012746  0.020351  0.012992  0.011741   \n",
       "Pan troglodytes                 0.012166  0.018944  0.013374  0.012893   \n",
       "Mus musculus                    0.013695  0.021015  0.011820  0.012234   \n",
       "Rattus norvegicus               0.012000  0.019189  0.011303  0.011862   \n",
       "Heterocephalus glaber           0.014855  0.021198  0.010091  0.009956   \n",
       "Danio rerio                     0.013429  0.018634  0.014449  0.010064   \n",
       "Gallus gallus                   0.011755  0.020823  0.012221  0.011894   \n",
       "Taeniopygia guttata             0.006985  0.037443  0.006003  0.011755   \n",
       "Xenopus tropicalis              0.014122  0.016669  0.014585  0.011962   \n",
       "Ambystoma mexicanum             0.009940  0.020192  0.011104  0.013472   \n",
       "\n",
       "                                     GGT       GGC       GGA       GGG  \n",
       "Species                                                                 \n",
       "Escherichia coli                0.024344  0.028417  0.008636  0.011247  \n",
       "Bacillus subtilis               0.012549  0.024139  0.021719  0.011159  \n",
       "Natronobacterium gregoryi SP2   0.014192  0.038010  0.012249  0.015762  \n",
       "Streptomyces coelicolor A3(2)   0.009214  0.061030  0.007141  0.018068  \n",
       "Deinococcus radiodurans R1      0.007219  0.058150  0.005437  0.020614  \n",
       "Azotobacter vinelandii          0.012301  0.069409  0.004380  0.008922  \n",
       "Mycoplasma genitalium           0.021639  0.008746  0.014465  0.018176  \n",
       "Acidianus brierleyi             0.018843  0.006470  0.033579  0.005159  \n",
       "Pyrolobus fumarii 1A            0.021228  0.031076  0.008383  0.014051  \n",
       "Pyrococcus abyssi GE5           0.015573  0.010826  0.030015  0.015971  \n",
       "Synechococcus lividus PCC 6715  0.015278  0.025661  0.008130  0.021440  \n",
       "Drosophila melanogaster         0.013375  0.026718  0.017880  0.004639  \n",
       "Caenorhabditis elegans          0.011753  0.006468  0.035210  0.003825  \n",
       "Tribolium castaneum             0.011648  0.017081  0.013171  0.013516  \n",
       "Felis catus                     0.009802  0.023086  0.016272  0.018202  \n",
       "Canis lupus familiaris          0.010706  0.023305  0.014515  0.017677  \n",
       "Arabidopsis thaliana            0.021563  0.008968  0.023635  0.010142  \n",
       "Aspergillus nidulans            0.020633  0.025152  0.014415  0.010541  \n",
       "Saccharomyces cerevisiae        0.025169  0.010006  0.010591  0.006111  \n",
       "Schizosaccharomyces pombe       0.021596  0.008316  0.015847  0.004383  \n",
       "Zea mays                        0.013636  0.032336  0.013510  0.017021  \n",
       "Chlamydomonas reinhardtii       0.010203  0.063039  0.005587  0.011586  \n",
       "Oryza sativa                    0.014947  0.026761  0.014789  0.015607  \n",
       "Lotus japonicus                 0.020569  0.011238  0.021229  0.012569  \n",
       "Dunaliella salina               0.015233  0.041546  0.009803  0.010685  \n",
       "Physcomitrella patens           0.018452  0.017262  0.022077  0.015831  \n",
       "Emiliania huxleyi               0.011556  0.049620  0.007716  0.012054  \n",
       "Thalassiosira pseudonana        0.022511  0.012007  0.025322  0.007434  \n",
       "Populus trichocarpa             0.021371  0.012546  0.021989  0.012965  \n",
       "Homo sapiens                    0.010503  0.023176  0.016677  0.020049  \n",
       "Pan troglodytes                 0.010009  0.020086  0.015998  0.016705  \n",
       "Mus musculus                    0.012358  0.021618  0.017658  0.015905  \n",
       "Rattus norvegicus               0.011491  0.021718  0.016793  0.015602  \n",
       "Heterocephalus glaber           0.012541  0.022914  0.017394  0.015442  \n",
       "Danio rerio                     0.013902  0.017138  0.022041  0.010038  \n",
       "Gallus gallus                   0.010778  0.018616  0.016405  0.016262  \n",
       "Taeniopygia guttata             0.007131  0.021200  0.014811  0.020312  \n",
       "Xenopus tropicalis              0.011857  0.014430  0.020429  0.014061  \n",
       "Ambystoma mexicanum             0.011502  0.022036  0.016871  0.018418  \n",
       "\n",
       "[39 rows x 64 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_abundances = pandas.read_csv(\"../data/codons_abundances_diverse_species.csv\", index_col=\"Species\")\n",
    "# normalize\n",
    "codon_abundances[\"sum\"] = codon_abundances.sum(axis=1)\n",
    "codon_abundances = codon_abundances.loc[:,\"TTT\":\"GGG\"].div(codon_abundances[\"sum\"], axis=0)\n",
    "codon_abundances"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "expensive-evening",
   "metadata": {},
   "source": [
    "### Function to read a genetic code from file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eastern-maryland",
   "metadata": {},
   "source": [
    "Format of the file: Tab-delimited file with two columns, amino acid in the first one, codon in the second. First line is header."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "positive-singer",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_code(fileName):\n",
    "    code = {}\n",
    "    with open(fileName, 'r') as f:\n",
    "        lines = f.read().split(\"\\n\")\n",
    "        # remove the header\n",
    "        lines.pop(0)\n",
    "        for l in lines:\n",
    "            if l!=\"\":\n",
    "                splitLine = l.split(\"\\t\")\n",
    "                #print(splitLine)\n",
    "                code[splitLine[1]] = splitLine[0]\n",
    "    return code"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "inclusive-purple",
   "metadata": {},
   "source": [
    "### Function to generate all codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "official-participant",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gen_all_codons():\n",
    "    res = []\n",
    "    bases = ['A', 'C', 'G', 'T']\n",
    "    for c1 in bases:\n",
    "        for c2 in bases:\n",
    "            for c3 in bases:\n",
    "                res = res+[c1+c2+c3]\n",
    "    return res"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "valued-nepal",
   "metadata": {},
   "source": [
    "## Generators of random codes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "august-manufacturer",
   "metadata": {},
   "source": [
    "Below we define ten different methods of generating alternative codes."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "electrical-innocent",
   "metadata": {},
   "source": [
    "### Amino acid permutation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "geological-gravity",
   "metadata": {},
   "source": [
    "Maintains the block structure of the standard genetic code, but changes which codons correspond to which amino acid. Position of the stop codons is fixed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "incorrect-tackle",
   "metadata": {},
   "outputs": [],
   "source": [
    "# standard_code ... dictionary containing the standard code; keys = codons, values = amino acids\n",
    "def rand_aa_permutation(standard_code, seed = 0):\n",
    "    np.random.seed(seed)\n",
    "    # all amino acids\n",
    "    aas = ['A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y', '*']\n",
    "    # dictionary to store the code\n",
    "    code = {}\n",
    "    # randomly shuffle the amino acids \n",
    "    shuffled_aas = np.random.permutation(20)\n",
    "    # the last one is always stop\n",
    "    shuffled_aas = np.append(shuffled_aas, [20])\n",
    "    # go over the standard code and always change amino acid i to shuffled_aas[i]\n",
    "    for codon, aa in standard_code.items():\n",
    "        code[codon] = aas[shuffled_aas[aas.index(aa)]]\n",
    "    return code"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "precious-spanish",
   "metadata": {},
   "source": [
    "### Amino acid permutation with restriction on the change in number of codons per amino acid"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "identified-bonus",
   "metadata": {},
   "source": [
    "Amino acid frequency in proteins is positively correlated with number of codons of the amino acid (Gilis, Genome Biology, 2001). This function generates randomized codes using amino acid permutation, but only permutations where number of codons of each amino acid changes by at most `max_diff` are allowed. Position of the stop codons is fixed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "independent-diabetes",
   "metadata": {},
   "outputs": [],
   "source": [
    "# standard_code ... dictionary containing the standard code; keys = codons, values = amino acids\n",
    "# max_diff ... maximal allowed change in number of codons per amino acid\n",
    "def rand_aa_permutation_restricted(standard_code, max_diff, seed = 0):\n",
    "    np.random.seed(seed)\n",
    "\n",
    "    # count the number of codons for each amino acid in the standard code\n",
    "    old_aas, counts = np.unique(list(standard_code.values()), return_counts=True)\n",
    "    old_aas = old_aas.tolist()\n",
    "    counts = counts.tolist()\n",
    "    # delete entries corresponding to stop codons, as they are not permuted\n",
    "    stop_index = old_aas.index('*')\n",
    "    del old_aas[stop_index]\n",
    "    del counts[stop_index]\n",
    "    # an array to store the new permutation of amino acids\n",
    "    new_aas = ['']*20\n",
    "    # an array indicating whether an amino acid has been assigned a codon block already\n",
    "    used = [False]*20\n",
    "\n",
    "    # iterate over blocks of given size, from the largest to the smallest\n",
    "    for blockSize in [6,4,3,2,1]:\n",
    "        # choose all blocks of given size and shuffle them\n",
    "        blocks = [x for x in range(len(counts)) if counts[x]==blockSize]\n",
    "        np.random.shuffle(blocks)\n",
    "        # amino acids that need to be placed in this step (otherwise their number of codons would change to much)\n",
    "        must_have_pos = [old_aas[x] for x in range(len(old_aas)) if counts[x]==blockSize+max_diff and not used[x]]\n",
    "        # place those amino acids to the first len(must_have_pos) blocks and mark them as used\n",
    "        for i in range(len(must_have_pos)):\n",
    "            aa = must_have_pos[i]\n",
    "            new_aas[blocks[i]] = aa\n",
    "            used[old_aas.index(aa)] = True\n",
    "        # amino acids that may get a position now\n",
    "        may_have_pos = [old_aas[x] for x in range(len(old_aas)) if counts[x]>=blockSize-max_diff and counts[x]<blockSize+max_diff and not used[x]]\n",
    "        # shuffle the potential amino acids and fill in the remaining blocks\n",
    "        np.random.shuffle(may_have_pos)\n",
    "        for i in range(len(must_have_pos), len(blocks)):\n",
    "            aa = may_have_pos[i-len(must_have_pos)]\n",
    "            new_aas[blocks[i]] = aa\n",
    "            used[old_aas.index(aa)] = True\n",
    "\n",
    "    # construct the new code\n",
    "    old_aas += ['*']\n",
    "    new_aas += ['*']\n",
    "    code = {}\n",
    "    for codon, aa in standard_code.items():\n",
    "        code[codon] = new_aas[old_aas.index(aa)]\n",
    "\n",
    "    return code\t"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "manual-sellers",
   "metadata": {},
   "source": [
    "### Quartet shuffling"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "intimate-impossible",
   "metadata": {},
   "source": [
    "Randomizes the first two positions of codons, while maintaining that the two sets of first and second positions in which the stop codons reside are separated by a single transition mutation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "beautiful-novel",
   "metadata": {},
   "outputs": [],
   "source": [
    "# standard_code ... dictionary containing the standard code; keys = codons, values = amino acids\n",
    "def rand_quartet_shuffling(standard_code, seed = 0):\n",
    "    np.random.seed(seed)\n",
    "    code = standard_code.copy()\n",
    "    bases = ['A', 'C', 'G', 'T']\n",
    "    # all possible pairs of bases\n",
    "    pairs = ['TA', 'TG', 'TT', 'TC', 'CT', 'CC', 'CA', 'CG', 'AT', 'AC', 'AA', 'AG', 'GT', 'GC', 'GA', 'GG']\n",
    "    # permutation of the pairs to construct the alternative codon\n",
    "    perm_pairs = pairs.copy()\n",
    "    np.random.shuffle(perm_pairs)\n",
    "    \n",
    "    # the \"opal\" stop codon is separated by a transition from the \"ambre\" stop codon\n",
    "    stop_pair_1 = perm_pairs[0]\n",
    "    # the transition can happen in the first, or in the second position\n",
    "    transition_pos = np.random.randint(2)\n",
    "    if transition_pos==0:\n",
    "        stop_pair_2 = transitions[stop_pair_1[0]]+stop_pair_1[1]\n",
    "    else:\n",
    "        stop_pair_2 = stop_pair_1[0]+transitions[stop_pair_1[1]]\n",
    "    tmp_pos = perm_pairs.index(stop_pair_2)\n",
    "    perm_pairs[tmp_pos] = perm_pairs[1]\n",
    "    perm_pairs[1] = stop_pair_2\n",
    "    # construct the new code\n",
    "    for i in range(len(pairs)):   \n",
    "        for b in bases:\n",
    "            old_codon = pairs[i]+b\n",
    "            new_codon = perm_pairs[i]+b\n",
    "            code[new_codon] = standard_code[old_codon]\n",
    "                                            \n",
    "    return code"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "statutory-float",
   "metadata": {},
   "source": [
    "### N-block Shuffler, Caporaso, J Mol Evol, 2005"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "italian-skating",
   "metadata": {},
   "source": [
    "\"The N-Blocks of the genetic code are held constant. We define an N-Block as a 1-, 2-, or 4-codon block in which all codons specify the same amino acid. The block structure is defined based on the canonical code, and all blocks of the same size are permuted among themselves. (...) The N-Block Shuffler maintains the number of codons per amino acid and the block structure of the genetic code.\" The position of stop codons is fixed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "agreed-return",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rand_N_block_shuffler(seed = 0):\n",
    "    np.random.seed(seed)\n",
    "    \n",
    "    # the block structure of the standard code\n",
    "    N_blocks_1 = ['ATA', 'ATG', 'TGG']\n",
    "    N_blocks_1_aas = ['I', 'M', 'W']\n",
    "    N_blocks_2 = [['TTT', 'TTC'], ['TTA', 'TTG'], ['TAT', 'TAC'], ['TGT', 'TGC'], ['CAT', 'CAC'], ['CAA', 'CAG'],\n",
    "        ['ATT', 'ATC'], ['AAT', 'AAC'], ['AAA', 'AAG'], ['AGT', 'AGC'], ['AGA', 'AGG'], ['GAT', 'GAC'], ['GAA', 'GAG']]\n",
    "    N_blocks_2_aas = ['F', 'L', 'Y', 'C', 'H', 'Q', 'I', 'N', 'K', 'S', 'R', 'D', 'E']\n",
    "    N_blocks_4 = [['TCT', 'TCC', 'TCA', 'TCG'], ['CTT', 'CTC', 'CTA', 'CTG'], ['CCT', 'CCC', 'CCA', 'CCG'], \n",
    "        ['CGT', 'CGC', 'CGA', 'CGG'], ['ACT', 'ACC', 'ACA', 'ACG'], ['GTT', 'GTC', 'GTA', 'GTG'], ['GCT', 'GCC', 'GCA', 'GCG'],\n",
    "        ['GGT', 'GGC', 'GGA', 'GGG']]\n",
    "    N_blocks_4_aas = ['S', 'L', 'P', 'R', 'T', 'V', 'A', 'G']\n",
    "    stop_codons = ['TAA', 'TAG', 'TGA']\n",
    "    \n",
    "    code = {}\n",
    "    # shuffle 1-blocks\n",
    "    perm1 = np.random.permutation(len(N_blocks_1))\n",
    "    for i in range(len(N_blocks_1)):\n",
    "        code[N_blocks_1[i]] = N_blocks_1_aas[perm1[i]]\n",
    "    # shuffle 2-blocks\n",
    "    perm2 = np.random.permutation(len(N_blocks_2))\n",
    "    for i in range(len(N_blocks_2)):\n",
    "        for codon in N_blocks_2[i]:\n",
    "            code[codon] = N_blocks_2_aas[perm2[i]]\n",
    "    # shuffle 4-blocks\n",
    "    perm4 = np.random.permutation(len(N_blocks_4))\n",
    "    for i in range(len(N_blocks_4)):\n",
    "        for codon in N_blocks_4[i]:\n",
    "            code[codon] = N_blocks_4_aas[perm4[i]]\n",
    "    # add stop codons\n",
    "    for codon in stop_codons:\n",
    "        code[codon] = \"*\"\n",
    "\n",
    "    return code"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "romantic-emergency",
   "metadata": {},
   "source": [
    "### Codon Shuffler, Caporaso, J Mol Evol, 2005"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "appropriate-listing",
   "metadata": {},
   "source": [
    "Amino acids are assigned to codons randomly, but the number of codons each amino acid has is preserved.\n",
    "Maintains the number of codons per amino acid, but produces codes without block structure. Position of stop codons fixed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "electrical-month",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rand_codon_shuffler(standard_code, seed = 0):\n",
    "    np.random.seed(seed)\n",
    "    # dictionary to store the code\n",
    "    code = standard_code.copy()\n",
    "    # count the number of codons for each amino acid\n",
    "    values, counts = np.unique(list(standard_code.values()), return_counts=True)\n",
    "    # shuffle the codons\n",
    "    # stop codons are fixed - don't include them in the permutation\n",
    "    codons = list(standard_code.keys())\n",
    "    perm = np.random.permutation(64)\n",
    "    codons = [codons[x] for x in perm if codons[x] not in ['TAA', 'TAG', 'TGA']]\n",
    "    # assign the codons to amino acids\n",
    "    start = 0\n",
    "    for i in range(len(values)):\n",
    "        aa = values[i]\n",
    "        if aa!=\"*\":\n",
    "            count = counts[i]\n",
    "            for j in range(start, start+count):\n",
    "                code[codons[j]]=aa\n",
    "            start = start+count\n",
    "    return code"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sexual-puzzle",
   "metadata": {},
   "source": [
    "### AAAGALOC Shuffler, Caporaso, J Mol Evol, 2005"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "magnetic-satisfaction",
   "metadata": {},
   "source": [
    "Generates completely random codes, the only restriction is that each amino acid has at least one codon. The resulting codes do not in general have a block structure. Number of codons per amino acid is not preserved. Position of stop codons fixed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "normal-alarm",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rand_aaagaloc_shuffler(standard_code, seed = 0):\n",
    "    np.random.seed(seed)\n",
    "    # dictionary to store the code\n",
    "    code = {}\n",
    "    # list of amino acids\n",
    "    aas = ['A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y']\n",
    "    # fix stop codons\n",
    "    code['TAA'] = \"*\"\n",
    "    code['TAG'] = \"*\"\n",
    "    code['TGA'] = \"*\"\n",
    "    # shuffle the codons\n",
    "    # stop codons are fixed - don't include them in the permutation\n",
    "    codons = list(standard_code.keys())\n",
    "    perm = np.random.permutation(64)\n",
    "    codons = [codons[x] for x in perm if codons[x] not in ['TAA', 'TAG', 'TGA']]\n",
    "    # the first 20 codons are assigned each to the corresponding amino acid\n",
    "    # to fulfill the requirement that each amino acid has at least one codon\n",
    "    for i in range(20):\n",
    "        code[codons[i]] = aas[i]\n",
    "    # the remaining codons are assigned to amino acids randomly\n",
    "    for i in range(41):\n",
    "        codon = codons[20+i]\n",
    "        aa = aas[np.random.randint(0,20)]\n",
    "        code[codon] = aa\n",
    "\n",
    "    return code"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "infrared-purse",
   "metadata": {},
   "source": [
    "### The Massey models, J Mol Evol, 2008"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "automatic-stake",
   "metadata": {},
   "source": [
    "The models by Massey (J Mol Evol, 2008) attempt to model the evolution of the genetic code, in particular the stepwise addition of new amino acids to the growing code.\n",
    "\n",
    "The amino acid to be added next is chosen randomly from all not yet placed amino acids with probability proportional to its physicochemical similarity to the parent acid. The physicochemical similarity is defined as 1/*d*, where *d* is the Grantham distance of the proposed amino acid and the parent amino acid.\n",
    "\n",
    "These models preserve the block structure of the genetic code. Further, the models tend to create codes in which neighbouring amino acids are physicochemically similar."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "interesting-dover",
   "metadata": {},
   "source": [
    "#### Random expansion"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dangerous-sierra",
   "metadata": {},
   "source": [
    "In Random expansion (Model 1 in Massey, J Mol Evol 2008), we start with a random amino acid assigned to a random codon block. Subsequent amino acids are chosen based on their similarity with the amino acid previously added to the code and placed in a randomly chosen codon block neighbouring the previous one. If no such codon block exists, a codon block is chosen randomly from the remaining free blocks."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "likely-export",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Helper function to get all neighbouring codon blocks of a given block.\n",
    "bl ... a codon block; list of codons\n",
    "blocks ... a list of blocks, each list of codons\n",
    "'''\n",
    "def get_neighbouring_blocks(bl, blocks):\n",
    "    neighbours = []\n",
    "    for codon in bl:\n",
    "        neighbours.append(gen_neighbouring_codons_ti(codon)+gen_neighbouring_codons_tv(codon))\n",
    "    neighbours = np.unique(neighbours)\n",
    "    # return blocks that share at least one element with neighbours\n",
    "    return [b for b in blocks if set(b) & set(neighbours)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "grave-citizenship",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rand_random_expansion(seed = 0):\n",
    "    np.random.seed(seed)\n",
    "    # dictionary to store the code\n",
    "    code = {}\n",
    "    # define the blocks of the code\n",
    "    blocks = [[\"TTT\", \"TTC\"], [\"TAT\", \"TAC\"], [\"ATA\", \"ATC\", \"ATT\"], [\"AAT\", \"AAC\"], \n",
    "        [\"AGT\", \"AGC\", \"TCA\", \"TCC\", \"TCG\", \"TCT\"], [\"ACA\", \"ACC\", \"ACG\", \"ACT\"], [\"TGT\", \"TGC\"], \n",
    "        [\"CGA\", \"CGC\", \"CGG\", \"CGT\", \"AGA\", \"AGG\"], [\"TTA\", \"TTG\", \"CTA\", \"CTC\", \"CTG\", \"CTT\"],\n",
    "        [\"CCA\", \"CCC\", \"CCT\", \"CCG\"], [\"CAT\", \"CAC\"], [\"GTG\", \"GTT\", \"GTA\", \"GTC\"], \n",
    "        [\"GCA\", \"GCG\", \"GCC\", \"GCT\"], [\"GAT\", \"GAC\"], [\"GGA\", \"GGC\", \"GGG\", \"GGT\"], \n",
    "        [\"TGG\"], [\"CAA\", \"CAG\"], [\"ATG\"], [\"AAA\", \"AAG\"], [\"GAA\", \"GAG\"]]\n",
    "    # list of aminoacids \n",
    "    aminoacids = ['A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y']\n",
    "\n",
    "    previous_block = None\n",
    "    previous_aa = None\n",
    "    for i in range(20):\n",
    "        # the first amino acid and block\n",
    "        if previous_aa == None:\n",
    "            # choose random block and random amino acid\n",
    "            next_block = blocks[np.random.randint(len(blocks))]\n",
    "            next_aa = aminoacids[np.random.randint(len(aminoacids))]\n",
    "        else:\n",
    "            # choose a block neighbouring the previous one\n",
    "            candidate_blocks = get_neighbouring_blocks(previous_block, blocks)\n",
    "            # at least one neighbouring block is free\n",
    "            if len(candidate_blocks)>0:\n",
    "                # choose a random one from the neighbouring blocks\n",
    "                next_block = candidate_blocks[np.random.randint(len(candidate_blocks))]\n",
    "            # there are no free neighbouring blocks\n",
    "            else:\n",
    "                # choose a random block from all remaining free blocks\n",
    "                next_block = blocks[np.random.randint(len(blocks))]\n",
    "            # probabilities of adding the individual amino acids\n",
    "            probs = [1/grantham_matrix.at[previous_aa, aa] for aa in aminoacids]\n",
    "            probs = [x/sum(probs) for x in probs]\n",
    "            # choose the next amino acid\n",
    "            next_aa = aminoacids[np.random.choice(len(aminoacids), p=probs)]            \n",
    "        # update aminoacids and blocks    \n",
    "        del aminoacids[aminoacids.index(next_aa)]\n",
    "        del blocks[blocks.index(next_block)]\n",
    "        # add the new codons to the code\n",
    "        for codon in next_block:\n",
    "            code[codon] = next_aa\n",
    "        previous_block = next_block\n",
    "        previous_aa = next_aa  \n",
    "\n",
    "    # add stop codons\n",
    "    stop_codons = [\"TAA\", \"TAG\", \"TGA\"]\n",
    "    for stop in stop_codons:\n",
    "        code[stop] = \"*\"\n",
    "\n",
    "    return code"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "proprietary-gathering",
   "metadata": {},
   "source": [
    "#### Ambiguity reduction 1, Ambiguity reduction 2 and 2-1-3 model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "designing-parameter",
   "metadata": {},
   "source": [
    "These are Model 2a, 2b and 3 in Massey, J Mol Evol, 2008. The codon blocks are added in order specified by Fig. 2a, 2b and 3 in Massey, J Mol Evol, 2008. In addition, in Model 3 the four starting amino acids are fixed (V, A, D, and G)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "harmful-package",
   "metadata": {},
   "source": [
    "Below we define the order of adding blocks, as well as the parent of each newly added block (0 if no parent) for models 2a, 2b and the 213 model:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "allied-mention",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Model 2a\n",
    "blocks_massey_2a = [[\"TTT\", \"TTC\"], [\"TAT\", \"TAC\"], [\"ATA\", \"ATC\", \"ATT\"], [\"AAT\", \"AAC\"], \n",
    "        [\"AGT\", \"AGC\", \"TCA\", \"TCC\", \"TCG\", \"TCT\"], [\"ACA\", \"ACC\", \"ACG\", \"ACT\"], [\"TGT\", \"TGC\"], \n",
    "        [\"CGA\", \"CGC\", \"CGG\", \"CGT\", \"AGA\", \"AGG\"], [\"TTA\", \"TTG\", \"CTA\", \"CTC\", \"CTG\", \"CTT\"],\n",
    "        [\"CCA\", \"CCC\", \"CCT\", \"CCG\"], [\"CAT\", \"CAC\"], [\"GTG\", \"GTT\", \"GTA\", \"GTC\"], \n",
    "        [\"GCA\", \"GCG\", \"GCC\", \"GCT\"], [\"GAT\", \"GAC\"], [\"GGA\", \"GGC\", \"GGG\", \"GGT\"], \n",
    "        [\"TGG\"], [\"CAA\", \"CAG\"], [\"ATG\"], [\"AAA\", \"AAG\"], [\"GAA\", \"GAG\"]]\n",
    "parents_massey_2a = [0, 0, 1, 2, 1, 3, 2, 4, 1, 5, 2, 3, 6, 4, 0, 7, 11, 3, 4, 14]\n",
    "\n",
    "# Model 2b\n",
    "blocks_massey_2b = [[\"TTA\", \"TTG\", \"CTA\", \"CTC\", \"CTG\", \"CTT\"], [\"CAT\", \"CAC\"], [\"GTG\", \"GTT\", \"GTA\", \"GTC\"],\n",
    "    [\"GAT\", \"GAC\"], [\"CCA\", \"CCC\", \"CCT\", \"CCG\"], [\"CGA\", \"CGC\", \"CGG\", \"CGT\", \"AGA\", \"AGG\"], \n",
    "    [\"GCA\", \"GCG\", \"GCC\", \"GCT\"], [\"GGA\", \"GGC\", \"GGG\", \"GGT\"], [\"TTT\", \"TTC\"], \n",
    "    [\"AGT\", \"AGC\", \"TCA\", \"TCC\", \"TCG\", \"TCT\"], [\"TAT\", \"TAC\"], [\"TGT\", \"TGC\"], [\"ATA\", \"ATC\", \"ATT\"], \n",
    "    [\"ACA\", \"ACC\", \"ACG\", \"ACT\"], [\"AAT\", \"AAC\"], [\"TGG\"], [\"ATG\"], [\"CAA\", \"CAG\"], [\"AAA\", \"AAG\"], \n",
    "    [\"GAA\", \"GAG\"]]\n",
    "parents_massey_2b = [0, 0, 1, 2, 1, 2, 3, 4, 1, 5, 2, 6, 3, 7, 4, 12, 13, 2, 15, 4]\n",
    "\n",
    "# Model 3 (213 model)\n",
    "blocks_213 = [[\"GTG\", \"GTT\", \"GTA\", \"GTC\"], [\"GCA\", \"GCG\", \"GCC\", \"GCT\"], [\"GAT\", \"GAC\"],\n",
    "    [\"GGA\", \"GGC\", \"GGG\", \"GGT\"], [\"TTA\", \"TTG\", \"CTA\", \"CTC\", \"CTG\", \"CTT\"],  [\"CCA\", \"CCC\", \"CCT\", \"CCG\"],\n",
    "    [\"CAT\", \"CAC\"], [\"CGA\", \"CGC\", \"CGG\", \"CGT\", \"AGA\", \"AGG\"], [\"TTT\", \"TTC\"], \n",
    "    [\"AGT\", \"AGC\", \"TCA\", \"TCC\", \"TCG\", \"TCT\"], [\"TAT\", \"TAC\"], [\"TGT\", \"TGC\"], [\"ATA\", \"ATC\", \"ATT\"], \n",
    "    [\"ACA\", \"ACC\", \"ACG\", \"ACT\"], [\"AAT\", \"AAC\"], [\"TGG\"], [\"CAA\", \"CAG\"], [\"ATG\"], [\"AAA\", \"AAG\"], \n",
    "    [\"GAA\", \"GAG\"]]\n",
    "parents_213 = [0, 0, 0, 0, 1, 2, 3, 4, 5, 6, 7, 8, 1, 2, 3, 12, 7, 13, 15, 3]\n",
    "fixed_aas_213 = {1: 'V', 2: 'A', 3: 'D', 4: 'G'}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "copyrighted-adobe",
   "metadata": {},
   "source": [
    "The function to generate codes according to models 2a, 2b or 3:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "buried-ireland",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "blocks ... the codon blocks, in the order in which they are added to the code; list of lists of codons\n",
    "parents ... a list of the same length as blocks, specifying the parent for each block in blocks; \n",
    "    0 if a block has no parent\n",
    "fixed_aas ... a dictionary in the form index:aa, index is index of a block, aa the amino acid that will be\n",
    "    assigned to this block; None if not applicable (no blocks with fixed meaning)\n",
    "'''\n",
    "def rand_massey_2_3_shuffler(blocks, parents, fixed_aas = None, seed = 0):\n",
    "    np.random.seed(seed)\n",
    "    # list of amino acids\n",
    "    aminoacids = ['A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y']\n",
    "    # list to store the amino acids added to the code\n",
    "    new_aas = []\n",
    "    # a dictionary to store the code\n",
    "    code = {}\n",
    "    for i in range(len(blocks)):\n",
    "        # no parent: either use the specified fixed aas or choose a random amino acid\n",
    "        if parents[i]==0:\n",
    "            if fixed_aas == None:\n",
    "                aa = aminoacids[np.random.randint(len(aminoacids))]\n",
    "            else:\n",
    "                aa = fixed_aas[i+1]\n",
    "        # choose amino acid with probability proportional to 1/d(parent, aa), d the Grantham distance\n",
    "        else:\n",
    "            parent_aa = new_aas[parents[i]-1]\n",
    "            # the probabilities of choosing the individual amino acids\n",
    "            probs = [1/grantham_matrix.at[parent_aa, aa] for aa in aminoacids]\n",
    "            probs = [x/sum(probs) for x in probs]\n",
    "            # choose a random amino acid with probability proportional to probs\n",
    "            aa = aminoacids[np.random.choice(len(aminoacids), p=probs)]\n",
    "        # update new_aas, delete the newly added amino acid from aminoacids\n",
    "        new_aas.append(aa)\n",
    "        del aminoacids[aminoacids.index(aa)]\n",
    "        # update the code\n",
    "        for codon in blocks[i]:\n",
    "            code[codon] = aa\n",
    "\n",
    "    # add stop codons\n",
    "    stop_codons = [\"TAA\", \"TAG\", \"TGA\"]\n",
    "    for stop in stop_codons:\n",
    "        code[stop] = \"*\"\n",
    "\n",
    "    return code"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "southwest-invalid",
   "metadata": {},
   "source": [
    "## ERMC"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "threatened-circular",
   "metadata": {},
   "source": [
    "The following function computes the expected random mutation cost (ERMC), as defined by Shehav and Zeevi, Science, 2020.\n",
    "\n",
    "Parameters:\n",
    "* code ... the genetic code; dictionary, keys = codons, values = amino acids\n",
    "* property_dict ... a dictionary with a chosen property of amino acids; keys = amino acids, values = values of the property; e.g. the `nitrogen_atoms`, `oxygen_atoms`, `polar_requirement`, `hydropathy` or `mol_volume` dictionaries defined above\n",
    "* mutRates ... a list of length two, the first item gives the mutation rate of a transition, the second item the mutation rate of a transversion\n",
    "* amino_acid_abundances ... a dictionary containing the median codon frequency of all codons encoding given amino acid; keys = amino acids, values = the medians\n",
    "* abs_val ... specifies if ERMC should be computed using absolute values of the differences in the property between the starting and ending amino acid (if abs_val==True), or if only positive differences should be taken into account (if abs_val==False); default False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "characteristic-prototype",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ERMC(code, property_dict, mutRates, amino_acid_abundances, abs_val=False):\n",
    "    ermc = 0\n",
    "    \n",
    "    codons = list(code.keys())\n",
    "\n",
    "    for c in codons:\n",
    "        aa = code[c]\n",
    "        freq = amino_acid_abundances[aa]\n",
    "        if aa!=\"*\":\n",
    "            neighs = gen_neighbouring_codons(c)\n",
    "            for n in neighs:\n",
    "                n_aa = code[n]\n",
    "                if n_aa!=\"*\":\n",
    "                    # mutation type\n",
    "                    mutType = get_mutation_type(c,n)\n",
    "                    # use absolute values in the computation of ERMC (used for mol. volume, PR, hydropathy)\n",
    "                    if abs_val == True:\n",
    "                        ermc += freq*mutRates[mutType] * abs(property_dict[n_aa] - property_dict[aa])\n",
    "                    else:\n",
    "                        ermc += freq*mutRates[mutType] * max(0, property_dict[n_aa] - property_dict[aa])\n",
    "                    \n",
    "    return ermc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "basic-murray",
   "metadata": {},
   "source": [
    "Function to compute all of the following:\n",
    "* ERMC in carbon conservation\n",
    "* ERMC in nitrogen conservation\n",
    "* ERMC in hydropathy\n",
    "* ERMC in polar requirement\n",
    "* ERMC in molecular volume\n",
    "for a given genetic code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "acceptable-latitude",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_code_characteristics(code, mutRates, codons_frequencies):\n",
    "    ermc_c = ERMC(code, carbon_atoms, mutRates, codons_frequencies, False)\n",
    "    ermc_n = ERMC(code, nitrogen_atoms, mutRates, codons_frequencies, False)\n",
    "    ermc_hydropathy = ERMC(code, hydropathy, mutRates, codons_frequencies, True)\n",
    "    ermc_polarreq = ERMC(code, polar_requirement, mutRates, codons_frequencies, True)\n",
    "    ermc_volume = ERMC(code, mol_volume, mutRates, codons_frequencies, True)\n",
    "    return [ermc_c, ermc_n, ermc_hydropathy, ermc_polarreq, ermc_volume]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "crazy-trick",
   "metadata": {},
   "source": [
    "## Plotting"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "introductory-building",
   "metadata": {},
   "source": [
    "Below we provide a function to plot the results. \n",
    "\n",
    "Parameters:\n",
    "* res ... the numpy array with results from randomized codes, as generated below\n",
    "* variable1 ... the plotted variable; one from \"nitrogen\", \"carbon\", \"hydropathy\", \"PR\", \"volume\"\n",
    "* variable2 ... if specified, only codes performing at least as good as the standard code in variable2 will be chosen for plotting variable1; one from \"nitrogen\", \"carbon\", \"hydropathy\", \"PR\", \"volume\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "connected-savannah",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_code_histograms(res, variable1, variable2 = None):\n",
    "    indexes = {\"nitrogen\": 1, \"carbon\": 0, \"hydropathy\": 2, \"PR\": 3, \"volume\": 4}\n",
    "    colors = {\"nitrogen\": '#0d4c7c', \"carbon\": '#151515', \"hydropathy\": '#018571', \"PR\":  '#660099', \"volume\": \"#330000\"}\n",
    "   \n",
    "    if variable2 == None:\n",
    "        results = res[:,indexes[variable1]]\n",
    "    else: \n",
    "        results = res[:,indexes[variable1]][res[:,indexes[variable2]]<=res[0,indexes[variable2]]] \n",
    "                          \n",
    "    color = colors[variable1]\n",
    "    \n",
    "    _, ax = plt.subplots(1, figsize =(3.5,2.333), dpi=144)\n",
    "    ax.hist(results, color=color, bins=50, density=True)\n",
    "    ax.axvline(results[0], color='red', lw=1)\n",
    "    ax.axvline(results[0], color='k', lw=0.6)\n",
    "    if variable2 == None:\n",
    "        print('{}, p={}'.format(variable1, sum(results<=results[0])/len(results)))\n",
    "    else:\n",
    "        print('{} confounded by {}, n={}, p={}'.format(variable1, variable2, len(results), sum(results<=results[0])/len(results)))\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "buried-madison",
   "metadata": {},
   "source": [
    "# Main"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "japanese-truth",
   "metadata": {},
   "source": [
    "This part of the code performs the actual analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "impressed-weekend",
   "metadata": {},
   "source": [
    "**SPECIFY THE PARAMETERS HERE:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "prompt-hypothetical",
   "metadata": {},
   "outputs": [],
   "source": [
    "# number of codes to generate\n",
    "num_codes = 10000\n",
    "\n",
    "# randomization type; one of: aa_permutation, aa_permutation_restricted, quartet, nblock, codon, aaagaloc, \n",
    "#     random_expansion, ambiguity_reduction_1, ambiguity_reduction_2, 213\n",
    "# see above for description of the individual randomization schemes\n",
    "randomization_type = \"aa_permutation\"\n",
    "\n",
    "# maximum difference between number of codons per aa in the alternative codes and the standard one\n",
    "# only used when randomization_type = \"aa_permutation_restricted\"\n",
    "max_diff = 2\n",
    "\n",
    "# transition:transversion ratio\n",
    "ti_tv_ratio = 0.5\n",
    "\n",
    "# organism from which codon frequencies should be used (one of the species in `codon_abundances` - see above)\n",
    "# \"baseline\" if all codon frequencies should be equal\n",
    "# \"ocean\" for the ocean codon abundances and mutation rates (if organism is \"ocean\", the above specified transition:transversion ratio is ignored)\n",
    "organism = 'ocean'\n",
    "\n",
    "# number of characteristics of the code; default 5 (carbon conservation, nitrogen conservation, hydropathy, PR, volume)\n",
    "# ONLY CHANGE THIS IF YOU ADD OR REMOVE SOME OF THE CODE CHARACTERISTICS\n",
    "num_char = 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "treated-sociology",
   "metadata": {},
   "source": [
    "Note: The computation takes a few seconds for 10,000 alternative codes. Please be patient if requiring larger number of codes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "broken-kidney",
   "metadata": {},
   "outputs": [],
   "source": [
    "# codon frequencies\n",
    "if organism == \"baseline\":\n",
    "    codons_frequencies = {}\n",
    "    codons = gen_all_codons()\n",
    "    for c in codons:\n",
    "        codons_frequencies[c] = 1/64\n",
    "elif organism == \"ocean\":\n",
    "    with open(\"../data/codonabuns_ocean.csv\", 'r') as cf:\n",
    "        cf_lines = cf.read().split(\"\\n\")[1:]\n",
    "        codons_frequencies={}\n",
    "        for item in cf_lines:\n",
    "            splitItem = item.split(\",\")\n",
    "            codons_frequencies[splitItem[0]] = float(splitItem[1])\n",
    "else:\n",
    "    codons_frequencies = codon_abundances.loc[organism].to_dict()\n",
    "    \n",
    "# frequencies of codons encoding each amino acid\n",
    "standard_code = read_code(\"../data/code_standard.tsv\")\n",
    "amino_acid_abundances = {}   # dictionary, keys = amino acids, values = list of frequencies of codons encoding given amino acid\n",
    "for codon, frequency in codons_frequencies.items():\n",
    "    aa = standard_code[codon]\n",
    "    if aa in amino_acid_abundances:\n",
    "        amino_acid_abundances[aa] += [frequency]\n",
    "    else:\n",
    "        amino_acid_abundances[aa] = [frequency]\n",
    "# compute the medians\n",
    "for aa, freqs in amino_acid_abundances.items():\n",
    "    amino_acid_abundances[aa] = median(freqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "novel-sympathy",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mutation rates\n",
    "mutRates = {}\n",
    "# compute mutation rates from transition:transversion ratio\n",
    "if organism != \"ocean\":\n",
    "    rate_tis = (2*ti_tv_ratio)/(1+2*ti_tv_ratio)\n",
    "    tmp_rates = [rate_tis, 1-rate_tis]\n",
    "    # normalize so that 4*rate_tis + 8*rate_tvs = 1 \n",
    "    # (there are 4 transitions and 8 transversions and all the rates should sum to 1)\n",
    "    tmp_rates = [x/(4*rate_tis + 8*(1-rate_tis)) for x in tmp_rates]\n",
    "    # transitions\n",
    "    for ti in [\"AG\", \"GA\", \"CT\", \"TC\"]:\n",
    "        mutRates[ti] = tmp_rates[0]\n",
    "    # transversions\n",
    "    for tv in [\"AC\", \"AT\", \"CA\", \"CG\", \"GC\", \"GT\", \"TA\", \"TG\"]:\n",
    "        mutRates[tv] = tmp_rates[1]\n",
    "# or use the ocean mutation abundances if organism is \"ocean\"\n",
    "else:\n",
    "    with open(\"../data/mutabuns_ocean.csv\", 'r') as f:\n",
    "        data = f.read().split()[1:]\n",
    "        for item in data:\n",
    "            splitData = item.split(\",\")\n",
    "            mutRates[splitData[0]] = float(splitData[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "reserved-precipitation",
   "metadata": {},
   "outputs": [],
   "source": [
    "# numpy array to store the results\n",
    "res = np.empty(shape=(num_codes+1, num_char))\n",
    "\n",
    "# characteristics of the standard code\n",
    "res[0,:] = compute_code_characteristics(standard_code, mutRates, amino_acid_abundances)\n",
    "\n",
    "# generate num_codes alternative codes and compute their characteristics\n",
    "for seed in range(num_codes):\n",
    "    if randomization_type==\"aa_permutation\":\n",
    "        code = rand_aa_permutation(standard_code, seed)\n",
    "    elif randomization_type==\"aa_permutation_restricted\":\n",
    "        code = rand_aa_permutation_restricted(standard_code, max_diff, seed)\n",
    "    elif randomization_type==\"quartet\":\n",
    "        code = rand_quartet_shuffling(standard_code, seed)\n",
    "    elif randomization_type==\"nblock\":\n",
    "        code = rand_N_block_shuffler(seed)\n",
    "    elif randomization_type==\"codon\":\n",
    "        code = rand_codon_shuffler(standard_code, seed)\n",
    "    elif randomization_type==\"aaagaloc\":\n",
    "        code = rand_aaagaloc_shuffler(standard_code, seed)\n",
    "    elif randomization_type==\"random_expansion\":\n",
    "        code = rand_random_expansion(seed)\n",
    "    elif randomization_type==\"ambiguity_reduction_1\":\n",
    "        code = rand_massey_2_3_shuffler(blocks_massey_2a, parents_massey_2a, None, seed)\n",
    "    elif randomization_type==\"ambiguity_reduction_2\":\n",
    "        code = rand_massey_2_3_shuffler(blocks_massey_2b, parents_massey_2b, None, seed)\n",
    "    elif randomization_type==\"213\":\n",
    "        code = rand_massey_2_3_shuffler(blocks_213, parents_213, fixed_aas_213, seed)\n",
    "    else:\n",
    "        print(\"Randomization type must be one of 'aa_permutation', 'aa_permutation_restricted', 'quartet', 'nblock', 'codon', 'aaagaloc', 'random_expansion', 'ambiguity_reduction_1', 'ambiguity_reduction_2', '213'.\")\n",
    "        quit()\n",
    "    # compute code characteristics and save\n",
    "    res[seed+1,:] = compute_code_characteristics(code, mutRates, amino_acid_abundances)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "personal-infrared",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nitrogen, p=0.0643935606439356\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAcsAAAE7CAYAAACykKpeAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Il7ecAAAACXBIWXMAABYlAAAWJQFJUiTwAAATJklEQVR4nO3dfbBtdV3H8fdHIUUDTNMcn0KZkFtayi1TMQFNIguk1BknK3tAe7DIQqupKNScyilC7cGZzDB7sNEGiyLAlFLRsu7VounycMWrGaEJwkW5KMi3P9Y6dDicfX73nrP3/u1z7vs1s+fHWb+91v5u1pz7Ob/18FupKiRJ0mT36l2AJEmLzrCUJKnBsJQkqcGwlCSpwbCUJKnBsJQkqcGwlCSpwbCUJKnBsJQkqcGwlCSpwbCUJKnBsJQkqeGQ3gXMQpKPAUcAezqXIklaHEcBe6vq0Qe64pYMS+CIww477IHbtm17YO9CtoLbd+7kCuDxj388hx56aO9yJGlddu3axb59+9a17lYNyz3btm174I4dO3rXsSVcl/Bw4OKLL+ZhD3tY73IkaV22b9/Ozp0796xnXc9ZSpLUYFhKktRgWEqS1GBYSpLUYFhKktRgWEqS1GBYSpLUYFhKktRgWEqS1GBYSpLUsFWnu5P22xGnnj2xb++Fr55jJZIWlSNLSZIaDEtJkhoMS0mSGgxLSZIaDEtJkhoMS0mSGgxLSZIaDEtJkhqmEpZJnpfkDUnel2RvkkryJ411nprkoiQ3JtmX5N+TvCzJvadRkyRJ0zKtGXx+CfgG4HPAJ4Fj13pzkucAfwncBvwFcCNwKvDbwPHA86dUlyRJGzatsPxphpDcDZwAXDbpjUmOAP4A+BJwYlX967j8bOA9wPOSvKCq3jal2qR1cyo8STClw7BVdVlVXVNVtR9vfx7wYOBtS0E5buM2hhEqwI9Noy5Jkqahx0Tqzxjbi1fpey9wK/DUJPepqi/MryxtZo4AJc1Sj7B87NhevbKjqu5I8jHg64DHALvW2lCSHRO61jxnqoPLWkE6q+0a0NLW0uPWkSPH9uYJ/UvLHzD7UiRJatvUz7Osqu2rLR9HnMfNuRxJ0hbVY2S5NHI8ckL/0vKbZl+KJEltPcLyqrE9ZmVHkkOARwN3ANfOsyhJkibpEZbvGdtTVul7OnA/4ANeCStJWhQ9wvIdwGeAFyT5xqWFSe4L/Or44+93qEuSpFVN5QKfJKcDp48/PnRsn5Lk/PG/P1NVLweoqr1JXswQmv+Q5G0M092dxnBbyTsYpsCTJGkhTOtq2CcAL1qx7DHjC+DjwMuXOqrqnUlOAH4ReC5wX4ap8n4GeP1+zgQkSdJcTCUsq+oc4JwDXOdy4NnT+HxJkmbJ51lKktSwqSclkBaVU+FJW4sjS0mSGgxLSZIaDEtJkhoMS0mSGgxLSZIaDEtJkhoMS0mSGgxLSZIaDEtJkhoMS0mSGgxLSZIaDEtJkhoMS0mSGgxLSZIafESXNo21HnslSbPkyFKSpAbDUpKkBsNSkqQGw1KSpAbDUpKkBsNSkqQGbx3RQvH2EEmLyJGlJEkNhqUkSQ2GpSRJDYalJEkNXuAjzdlaFzHtvfDVc6xE0v5yZClJUoMjS82dt4dI2my6jiyTfEeSS5N8Msm+JNcmeXuSp/SsS5Kk5bqFZZLfAP4GOA64GHgdsBN4DnB5ku/tVZskSct1OQyb5KHAy4FPAV9fVZ9e1ncS8B7gVcCf9KhPkqTleo0sv3r87H9eHpQAVXUZcAvw4B6FSZK0Uq+wvAb4IvCkJF+5vCPJ04HDgb/vUZgkSSt1OQxbVTcm+TngXOA/k7wTuAE4GjgNeBfwI63tJNkxoevYKZUqSVK/W0eq6rwke4A3Ay9e1rUbOH/l4VlJknrpeTXszwLvAM5nGFHeH9gOXAv8aZLXtrZRVdtXewFXzrB0SdJBpktYJjkR+A3gr6vqZ6rq2qq6tap2At8F/DdwVpLH9KhPkqTleo0sv3NsL1vZUVW3Ah9iqO2J8yxKkqTV9ArL+4ztpNtDlpZ/cQ61SJK0pl5h+b6xfUmShy/vSPLtwPHAbcAH5l2YJEkr9boa9h0M91F+K7AryQXA9cA2hkO0AX6+qm7oVJ8kSXfpdZ/lnUmeDbwUeAHDRT33A24ELgJeX1WX9qhNkqSVet5neTtw3viShA+GlhaVz7OUNgmDVOqn6/MsJUnaDAxLSZIaDEtJkhoMS0mSGgxLSZIaDEtJkhoMS0mSGgxLSZIaDEtJkhoMS0mSGgxLSZIaDEtJkhoMS0mSGgxLSZIafESXZmKtx0lJ0mbjyFKSpAbDUpKkBsNSkqQGw1KSpAbDUpKkBsNSkqQGw1KSpAbDUpKkBsNSkqQGw1KSpAbDUpKkBsNSkqQGw1KSpAbDUpKkhu5hmeSZSS5Icn2SLyS5LsklSZ7duzZJkqDz8yyTvBZ4BfBJ4K+BzwAPBrYDJwIXdStOkqRRt7BM8mKGoHwL8JKq+uKK/kO7FCZtQms9bHvvha+eYyXS1tTlMGyS+wCvAT7BKkEJUFW3z70wSZJW0Wtk+SyGw63nAXcm+Q7gccBtwIeq6oOd6pIk6R56heU3je1twIcZgvIuSd4LPK+q/netjSTZMaHr2A1XKEnSqNfVsA8Z21cABXwLcDjw9cClwNOBt/cpTZKku+s1slwK6TuA06pqz/jzFUm+C7gKOCHJU9Y6JFtV21dbPo44j5tivZKkg1ivkeVNY/vhZUEJQFXdClwy/vikOdYkSdKqeoXlVWN704T+z47tYbMvRZKktfUKy3cznKv82iSr1bB0wc/H5leSJEmr6xKWVfVx4ELgUcBPLe9LcjLwbQyjzovnXpwkSSv0nO7upcATgXPH+yw/DDwaOB34EnBGVd3crzxpa3B2H2njuoVlVX0yyXbgl4HTGG4X2csw4vy1qvpQr9okSVqu60Tq46QDPzm+JElaSN0f0SVJ0qIzLCVJauh6GFab21oXjkjSVuLIUpKkBsNSkqQGw1KSpAbDUpKkBsNSkqQGw1KSpAbDUpKkBsNSkqQGw1KSpAbDUpKkBsNSkqQGw1KSpAbDUpKkBsNSkqQGw1KSpAbDUpKkBsNSkqQGw1KSpAbDUpKkBsNSkqSGQ3oXoMV2xKlnc+X438e86LXc676Hd61HknpwZClJUoNhKUlSg2EpSVKDYSlJUoNhKUlSg2EpSVLDwoRlku9NUuPrjN71SJK0ZCHCMskjgd8BPte7FkmSVuoelkkC/BFwA/DGzuVIknQP3cMSOBN4BvCDwOc71yJJ0j10ne4uyTbg14HXVdV7kzyjZz2S7u6IU8+e2Lf3wlfPsRKpr25hmeQQ4K3AJ4BfWOc2dkzoOna9dUkHk7XCcCPrGqTaanqOLH8ZeCLwtKra17EOSZLW1CUsk3wzw2jyt6rqg+vdTlVtn7D9HcBx692uJEnLzf0Cn/Hw6x8DVwPrPwYkSdKc9Lga9suBY4BtwG3LJiIo4FfG9/zBuOy8DvVJknQ3PQ7DfgH4wwl9xzGcx3w/cBWw7kO0kiRNy9zDcryYZ9Xp7JKcwxCWb6mqN82zLkmSJlmESQkkSVpohqUkSQ0LFZZVdU5VxUOwkqRFslBhKUnSIuo6N6wWw0amPJOkg4EjS0mSGgxLSZIaDEtJkho8Zylp6nx8l7YaR5aSJDUYlpIkNRiWkiQ1GJaSJDUYlpIkNRiWkiQ1GJaSJDUYlpIkNRiWkiQ1GJaSJDUYlpIkNRiWkiQ1GJaSJDUYlpIkNRiWkiQ1GJaSJDX48GdJc+WDobUZObKUJKnBsJQkqcGwlCSpwbCUJKnBsJQkqcGwlCSpwVtHDgJrXaovSWrrMrJM8qAkZyS5IMnuJPuS3Jzk/Ul+OIkjXknSwug1snw+8PvA/wCXAZ8Avgr4buBNwLcneX5VVaf6JEm6S6+wvBo4DfjbqrpzaWGSXwA+BDyXITj/sk95kiT9vy6HO6vqPVV14fKgHJdfD7xx/PHEuRcmSdIqFvHc4O1je0fXKiRJGi3U1bBJDgG+f/zx4v14/44JXcdOrShJ0kFv0UaWvw48Drioqi7pXYwkSbBAI8skZwJnAVcC37c/61TV9gnb2gEcN73qJEkHs4UIyyQ/AbwO+E/gmVV1Y+eSJHXgsy61qLofhk3yMuANwH8AJ41XxEqStDC6jiyT/BzDecqPAM+qqs/0rEfS4nLUqZ66jSyTnM0QlDsYDr0alJKkhdRlZJnkRcCrgC8B7wPOTLLybXuq6vw5lyZJ0j30Ogz76LG9N/CyCe/5R+D8eRQjSdJauoRlVZ0DnNPjs7cqH8MlSbPT/WpYSZIWnWEpSVLDQkxKIEkb4W0lmjVHlpIkNRiWkiQ1GJaSJDUYlpIkNRiWkiQ1GJaSJDUYlpIkNRiWkiQ1OCnBJuL8r9KBc8ICTYMjS0mSGgxLSZIaDEtJkhoMS0mSGgxLSZIaDEtJkhq8dUTSQWsjt2N528nBxZGlJEkNhqUkSQ0ehl0wztIjSYvHkaUkSQ2GpSRJDYalJEkNnrOUpHXwaSYHF0eWkiQ1GJaSJDV4GLYDbw+RpM3FkaUkSQ2OLCVpyrz4Z+vpGpZJHgG8CjgFeBDwP8A7gVdW1Wc7liZJM2GQbk7dwjLJ0cAHgIcAfwVcCTwJ+CnglCTHV9UNveqTJGlJz5Hl7zEE5ZlV9YalhUnOBX4aeA3wo51q2y9eqCNpmhx1Lq4uYTmOKk8G9gC/u6L7V4CXAN+X5Kyq+vycy5OkhbPeP84N2enoNbI8aWwvrao7l3dU1S1JLmcI0ycD7553ccs5epQk9QrLx47t1RP6r2EIy2NYIyyT7JjQ9Q27du1i+/bt669w9PmPXrfhbWx2p4ztvn/+c7iXdxtJm8m9H/CmqW/zCUc/bOrbBPjIGv/eTuMzd+3aBXDUetbtFZZHju3NE/qXlj9gndv/0r59+27euXPnnnWuvz+OHdsrZ/gZC+GKsb3zlk93rWNBHDT7XXdxn6+wc+f1m/UzjwL2rmfFTX2fZVVtfOi4Tkuj2p41aP7c7wcf97mg3ww+SyPHIyf0Ly2/afalSJK0tl5hedXYHjOh/2vGdtI5TUmS5qZXWF42ticnuVsNSQ4HjgduBf5p3oVJkrRSl7Csqo8ClzKcbH3piu5XAvcH3uo9lpKkRdDzAp8fZ5ju7vVJngnsAr6Z4R7Mq4Ff7FibJEl3SVX1+/DkkdxzIvULcCJ1SdIC6RqWkiRtBk7HIklSg2EpSVKDYSlJUoNhKUlSg2EpSVKDYSlJUoNhuUySRyR5c5LrknwhyZ4k5yX5igPYxrOS/FaSdye5IUklef8s69b6bXSfJ7l/khcm+bMkVyb5fJJbkvxrkrOSfNmsv4MO3JR+11+R5KJx3c8l2ZvkiiTnJnnELOvX/Hmf5SjJ0QwzCj0E+CuGZ9c9iWFGoauA46vqhv3YzjuB5wC3AbuBxwGXV9XTZlO51msa+zzJKcDfATcyzHm8G/gK4DTgoeP2n1lVt83oa+gATfF3fTfwOeDfgE8BhwJPBE5geGbiiVX14Vl8B3VQVb6GPxguAQr4yRXLzx2Xv3E/t/MU4OuAezPMfVvA+3t/P1+z2efAE4AXAl+2YvnhwI5xO2f1/q6+prvfx/ffd8LyF4/buaj3d/U1vZcjS+76S3M3sAc4uqruXNZ3OMM0fAEeUgcwuXuSo4CP4chy4cxqn6/4jO8B/hT4m6o6dcNFa8PmtN+PZHgW7+6q+prG27VJeM5ycNLYXrr8lwegqm4BLgfuBzx53oVpZuaxz28f2zs2sA1N1zz2+9IfRv++gW1owRiWg8eO7aSHTV8ztpMeVq3NZx77/IfG9uINbEPTNfX9nuSMJOck+c0klwBvAT4O/Pz6y9Si6fmIrkVy5NjePKF/afkDZl+K5mSm+zzJTzA8TecjwJvXsw3NxCz2+xkMjxdc8i/A91TV7gMrTYvMkaU0ZUm+GzgPuB54blXdvvYa2syq6slVFeArgZPHxTuSfFvHsjRlhuVg6a/JIyf0Ly2/afalaE5mss+TnA68Dfg0w60D166nOM3MzH7Xq+qGqnoXQ2DuA96a5LADrlALybAcXDW2k85TLF3RNuk8hzafqe/zJM8H3s5wz90JVXVVYxXN38x/16vqJuCDwIMZbiPTFmBYDi4b25OT3O3/yXg5+fHArcA/zbswzcxU93mSFwJ/DlzHEJTXNFZRH/P6XX/42Hol9BZhWAJV9VHgUoZJBF66ovuVwP2Bty6/7yrJsUmOnVuRmqpp7vMkLwL+GPgE8HQPvS6uae33JI9K8lWrfUaSHwG+Cfgv4IrpVa+enJRgtMoUWLsYrnA7ieGQzFNr2RRYSQpgPLG/fDtPY7g6DuDLgecynL/6u6X3VNUPzOp7aP9NY58nOQn4e4Y/PN/M8A/kSjdV1Xmz+RY6UFPa76czHHL/IMMkB58CHsRwf+bjGabB+86q+sfZfyPNg2G5TJJHAq9iuOT/QQyzeVwAvLKqPrvivZPC8geAP1rrc1auo342us/3Z38DH6+qo6ZXtTZqCvv9UcCZwLcwjFIfyDAf9LXAu4DXVdVqfzhpkzIsJUlq8JylJEkNhqUkSQ2GpSRJDYalJEkNhqUkSQ2GpSRJDYalJEkNhqUkSQ2GpSRJDYalJEkNhqUkSQ2GpSRJDYalJEkNhqUkSQ2GpSRJDYalJEkNhqUkSQ3/B1QwpxmwIABZAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 504x335.952 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "carbon, p=0.0006999300069993001\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAb4AAAE7CAYAAACi8OSCAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Il7ecAAAACXBIWXMAABYlAAAWJQFJUiTwAAAR+0lEQVR4nO3df4xlZX3H8fdX3BpUQMEOBqVZf7DuRhZ0t1UrRkGTra2tpSqpf4Bif8VfpRJJmuIvNBJt2iqotSZaUNS20TRCGg1SkSL+atNZ6WpdwLYuliqMAoLSxR/Lt3+cM+12mDsze89z587M9/1KTg5zztzzPHcf5n7uc855nhOZiSRJVTxg2hWQJGk1GXySpFIMPklSKQafJKkUg0+SVIrBJ0kqxeCTJJVi8EmSSjH4JEmlGHySpFIMPklSKQafJKmUB067AsuJiG8CRwL7plwVSdLasRm4OzMfc6gvXPPBBxx5+OGHH71t27ajp12RhX6yezdfBbZv386mTZumXR1JKmPv3r3s379/rNeuh+Dbt23btqNnZ2enXY/7+XYEjwKuvPJKjjvuuGlXR5LK2LlzJ7t37943zmu9xidJKsXgkySVYvBJkkox+CRJpRh8kqRSDD5JUikGnySpFINPklSKwSdJKsXgkySVsh6mLJMmbmZmZuS+ubm5VayJpEmzxydJKsXgkySVYvBJkkox+CRJpRh8kqRSDD5JUikGnySpFMfxSctwjJ+0sRh82lAMKUnL8VSnJKkUg0+SVIrBJ0kqxeCTJJVi8EmSSjH4JEmlGHySpFIMPklSKQafJKkUg0+SVIpTlkkDOEWatP4YfCpjqZCSVIfBp3XHAJM0hNf4JEmlGHySpFIMPklSKRMLvog4MyKyX35nUuVIknQoJhJ8EXE88B7gh5M4viRJ42oefBERwKXA7cD7Wh9fkqQhJtHjOwd4NvAy4J4JHF+SpLE1HccXEduAtwMXZ+bnIuLZLY8vrSfO6iKtTc2CLyIeCHwY+BZw/hivnx2xa+uQekmSdLCWPb43Ak8GnpGZ+xseVwU5O4ukSWkSfBHxVLpe3p9l5pfGOUZm7hxx7Flgx4DqSZL0vwbf3NKf4rwMuAl4w+AaSZI0QS3u6nwosAXYBtx70KD1BN7U/877+20XNShPkqSxtTjV+SPgL0fs20F33e/zwI3AWKdBJUlqZXDw9TeyLDolWURcQBd8H8rMDwwtS5KkoZykWpJUisEnSSplosGXmRdkZniaU5K0VtjjkySVYvBJkkox+CRJpRh8kqRSDD5JUikGnySpFINPklSKwSdJKsXgkySVYvBJkkox+CRJpbR4Hp90yGZmZqZdBUlFGXzSFCwV/HNzc6tYE6keT3VKkkox+CRJpRh8kqRSDD5JUikGnySpFINPklSKwSdJKsVxfNIas9zgfsf5ScPY45MklWLwSZJKMfgkSaUYfJKkUgw+SVIpBp8kqRSDT5JUisEnSSrF4JMklWLwSZJKMfgkSaUYfJKkUpykWhOz3GTLkjQN9vgkSaUYfJKkUgw+SVIpBp8kqZQmwRcRfxwRV0fEf0bE/oi4IyK+EhFviohjWpQhSVILrXp85wIPAf4euBj4KPBT4AJgT0Qc36gcSZIGaTWc4cjMvHfhxoi4EDgf+CPglY3KkiRpbE16fIuFXu9j/fqEFuVIkjTUpG9u+bV+vWfC5UiStCJNZ26JiPOAhwJHAT8PPIMu9N6+gtfOjti1tVkFJUnltZ6y7Dzg2IN+vhI4OzO/27gcSZLG0jT4MvORABFxLPB0up7eVyLiVzNz9zKv3bnY9r4nuKNlPaX1bKk5UOfm5laxJtL6NJFrfJl5W2Z+AtgFHANcNolyJEk6VBO9uSUzbwa+DjwxIh4xybIkSVqJ1Ziy7Lh+fWAVypIkaUmDgy8itkTEUYtsf0A/gH0G+GJm3jm0LEmShmpxc8uvAG+LiM8D3wRup7uz81nAY4Fbgd9tUI4kSYO1CL7PAI+nG7P3ZOBhwD3ATcCHgXdl5h0NypEkabDBwZeZXwNe3aAukiRNnM/jkySV0nrmFhWz1GBqSVqL7PFJkkox+CRJpRh8kqRSDD5JUikGnySpFINPklSKwSdJKsXgkySVYvBJkkox+CRJpRh8kqRSDD5JUikGnySpFINPklSKwSdJKsXgkySVYvBJkkrxCexalk9Zl7SR2OOTJJVi8EmSSjH4JEmlGHySpFIMPklSKQafJKkUg0+SVIrj+KQNZKkxl3Nzc6tYE2ntsscnSSrF4JMklWLwSZJKMfgkSaUYfJKkUgw+SVIpBp8kqRSDT5JUisEnSSpl8MwtEXEM8BvA84DtwKOAHwNfBS4FLs3M+4aWI2kYZ3WROi2mLDsD+AvgO8A1wLeAY4EXAB8AfjkizsjMbFCWJEmDtAi+m4DnA588uGcXEecD/wS8kC4E/7ZBWZIkDTL4Gl9mfjYz/27h6czMvBV4X//jqUPLkSSphUk/neEn/fqnEy5H0gBe/1MlEwu+iHgg8JL+xytX8PuzI3ZtbVYpSVJ5kxzO8HbgROBTmfnpCZYjSdKKTaTHFxHnAK8FbgDOWslrMnPniGPNAjva1U6SVFnzHl9EvBq4GPg6cFpm3tG6DEmSxtU0+CLiNcC7ga/Rhd6tLY8vSdJQzYIvIv4QeCdwPV3oeSuYJGnNaRJ8EfEGuptZZoHnZOb3WhxXkqTWWszV+VLgLcAB4DrgnIhY+Gv7MvODQ8uSJGmoFnd1PqZfHwa8ZsTvXAt8sEFZkiQN0mLKsgsyM5ZZTm1QV0mSBvN5fJKkUiY9V6fWiaXmapSkjcQenySpFINPklSKwSdJKsXgkySVYvBJkkox+CRJpRh8kqRSDD5JUikGnySpFINPklSKwSdJKsXgkySV4iTVkpa01ATmc3Nzq1gTqQ17fJKkUgw+SVIpBp8kqRSDT5JUisEnSSrF4JMklWLwSZJKMfgkSaU4gF3S2BzcrvXIHp8kqRR7fIUs9e1ckqow+CRNhKdBtVZ5qlOSVIrBJ0kqxeCTJJVi8EmSSjH4JEmlGHySpFIMPklSKQafJKkUg0+SVIrBJ0kqpcmUZRHxIuBZwJOAk4EjgI9m5pktjq+Vcz5OSVpaq7k6X08XeD8EbgG2NjquJElNtTrVeS6wBTgSeEWjY0qS1FyTHl9mXjP/3xHR4pCSJE2EN7dIkkpZM8/ji4jZEbu8XihJasYenySplDXT48vMnYtt73uCO1a5OpKkDcoenySpFINPklSKwSdJKsXgkySV0mquztOB0/sfH9mvfzEiPtj/9/cy87wWZUmSNESruzqfBLx0wbbH9gvAzYDBJ0mauianOjPzgsyMJZbNLcqRJGkor/FJkkox+CRJpRh8kqRSDD5JUikGnySpFINPklTKmnk6g1ZuZmZm2lWQBhn3/+G5ubnGNVFF9vgkSaUYfJKkUgw+SVIpBp8kqRSDT5JUisEnSSrF4JMklWLwSZJKcQC7pHVjuYHvDnDXStjjkySVYvBJkkrxVOca5XyckjQZBp+kDWOpL4xe/9M8T3VKkkox+CRJpRh8kqRSDD5JUine3DJF3rkpSavPHp8kqRSDT5JUisEnSSrF4JMklWLwSZJK8a5OSSU4nZnm2eOTJJVi8EmSSvFUp6Tyxp1MwlOk65M9PklSKQafJKkUT3VOkHNxStLaY/BJ0pi8Nrg+NQu+iHg08BbgucAxwHeAy4E3Z+adrcpZi7Zv385hhx027WpIklagSfBFxOOALwIzwBXADcBTgD8AnhsRp2Tm7S3KkiRpiFY9vvfShd45mfnu+Y0R8Q7gXOBC4OWNypKkdc1ZZKZrcPD1vb1dwD7gzxfsfhPwe8BZEfHazLxnaHmStJENuSnO0FyZFj2+0/r1VZl538E7MvMHEfEFumB8GnB1g/Imwm9gklRDi+B7Qr++acT+b9AF3xaWCL6ImB2x6+S9e/eyc+fO8WvY27Nnz1iv27Rp06Lbt/Xru+66i4gYs1aS1Maoz6rlnHTSSSP3LfW5OYnXrdTevXsBNo/z2hbBd1S/vmvE/vntDxvz+Af2799/1+7du/eN+fpWtvbrG+Y3fLVfHzhwYPVro5W4X5tpzbPNpmD37t1DXnfIbTZueQtsBu4e54VrZhxfZg7v0k3QfI90rddT/8c2W39ss/VnPbZZiynL5nt0R43YP7/9+w3KkiRpkBbBd2O/3jJi/wn9etQ1QEmSVk2L4LumX++KiP93vIg4AjgF+G/gyw3KkiRpkMHBl5n/DlxFd6HxVQt2vxl4CPBhx/BJktaCVje3vJJuyrJ3RcRzgL3AU+nG+N0EvK5ROZIkDRKZ2eZAEcdz/0mqP0GBSaolSetHs+CTJGk98AnskqRSDD5JUikGnySpFINPklSKwSdJKsXgkySVUjr4IuLREXFJRHw7In4UEfsi4qKIePiAYz4zIg5EREbEW1vWV23bLCJ2RMRfRcQt/bFui4hrI+Ilk6h7Va3aLCKeERFX9K+/NyK+FRGfiojnTqruFUXEiyLi3RFxXUTc3X+WfWTMYzX/jG2h7Di+iHgc3WwzM8AVdM+SegrdbDM3Aqdk5u2HeMwjgD3AI4CHAhdm5utb1ruylm0WEa8GLgbuBD4J/BdwNHAicEtmvrj5GyioVZtFxCuA9wL30E2McQvwaOAFwIOB12fmhZN4D9VExPXAycAP6f6dtwIfzcwzD/E4zT9jm8nMkgvwaSCB31+w/R399veNccxLgDuA8/tjvHXa73MjLa3aDNgF3Ncf74hF9m+a9nvdKEuLNgM20T3WbD/whAX7tgH30k2E/6Bpv9+NsNAF0wlAAKf27fSRabT9pJaSPb7+m8i/AfuAx2XmfQftO4JuurUAZnKFk2tHxK8DlwNn0c2Bein2+Jpp2WYR8S/A44Gfy2l94yygVZtFxLHArcCezDx5kf17gO3AI2zPtiLiVLon8BxSj28Sn7EtVb3Gd1q/vurgBgHIzB8AX6A7ffK0lRwsImaA9wOXZ+ZY58K1rCZtFhEnAifRPVHkjog4LSLOi4jXRsRzFj5aS4O0+jubA74LbImIEw7eERFb6Hon1xt6a0rTz9jWqv6RP6Ffj3o47jf69aiH6y70frp/y5cPqZSW1KrNfqFfzwH/AHwW+BPgT4HPANdHxOPHr6YO0qTNsjst9Sq6v7HZiPhQRLwtIi4DZoF/Bc5oUF+10/oztqlWjyVab47q13eN2D+//WHLHSgifgt4PvCbmXnb8KpphFZtNtOvf5vuhpbnAZ8HjgXeCJwJfDIitmfmj8euraDh31lmfjwivg38NXDwXbe30V1W+I8x66jJaNb2k1C1x9dERGwGLgI+npkfm25ttELz/88fBrw4Mz+VmXdn5jfoPlD/me5b6AunVUHdX0ScSdcjv47uhpYH9+urgfcAfzO92mm9qRp88982jhqxf37795c5ziV0d5q9skGdtLRWbTa//9bM/NLBO/pTalf0Pz7lEOun+2vSZv11vEvoTmmelZk3ZOb+zLyB7mayWeCM/kYMrQ2t/l4nomrw3divR51fnr+APur89LwddKfOvtsP8syISLpTLwCv67ddPqi2gnZtNn+c74/YP//Q5MNXVi0toVWb7aIb0nDtIjdK3Ad8rv9x5ziV1ES0avuJqHqN75p+vSsiHrDIrban0I0L+vIyx7mM7pTLQicAzwSup/s2+pWhFVazNvsy3SDozRHxkEVupT6xX3+zQZ2ra9VmD+rXPzti//x2r8muHa3afjKmPVhyWguHOLiSbvaCrSs89tk4gH3NthndjC0JvJN+9qJ++3a6U9c/oRt7NPX3vN6XFm1Gd9o56T4oT1qw70l9m90HPHHa73ejLSwzgJ2uJ751sb+XQ2371VxKDmCHRafT2Qs8lW78yU3A0/OgcUH9KUwyM1Zw7LNxAHtzrdosIo4ErqX70PxHujFFx9JNf3U48JrMvHjCb6eEhm12CfAyul7dJ4Cbgc3A6cDPABdl5rmTfTc1RMTpdP+uAI8Efonurtnr+m3fy8zz+t/dTHd25ObM3LzgOIfU9qtq2t8opvxt5ni6gPoO3R/UzXR3aT58kd9N+vsfVnDcs7HHt6bbjH4uVbo/wB/RXfO7Ctg17fe40ZYWbUY3y8fZdGMv7wR+Sjc94NV0d+dO/X1ulAW4YL4dRiz7DvrdzQu3jdv2q7mU7fFJkmqqelenJKkog0+SVIrBJ0kqxeCTJJVi8EmSSjH4JEmlGHySpFIMPklSKQafJKkUg0+SVIrBJ0kqxeCTJJVi8EmSSjH4JEmlGHySpFIMPklSKQafJKmU/wECvjL4DyrL0QAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 504x335.952 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "volume, p=0.0032996700329967\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAd4AAAE7CAYAAABzF+DyAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Il7ecAAAACXBIWXMAABYlAAAWJQFJUiTwAAAWZ0lEQVR4nO3dfbBdVXnH8e/DSwUDhhelFHEawYSkiloSJRpLhYwxo4Ng1dbpNAWq01JxQMSpHa1CqFidqQiiaMcWM8WZUgdHsVMFZoCKvOkksb5eCG8XcaKiASKGQBWe/rH3wZPDOfeec88+695z8/3M3Fnc/bLWPtvt/WXtvfY6kZlIkqQy9pjtA5AkaXdi8EqSVJDBK0lSQQavJEkFGbySJBVk8EqSVJDBK0lSQQavJEkFGbySJBVk8EqSVJDBK0lSQQavJEkF7TXbBzDXRcS9wLOAyVk+FEnS3LEI+GVmPn/QHQ3e6T1r3333PWjZsmUHzfaBzGW/3ryZ7wFHH300e++992wfjiSN1MTEBDt37pzRvgbv9CaXLVt20KZNm2b7OOa0rRE8F7j66qs57LDDZvtwJGmkli9fzubNmydnsq/PeCVJKsjglSSpIINXkqSCDF5JkgoyeCVJKsjglSSpIINXkqSCDF5JkgoyeCVJKsjglSSpIKeMlLSLFREz3ndjZoNHIs1P9nglSSrI4JUkqaDGgjciDo+IyyJia0Q8HhGTEXFRRBw4QB2viYiPRcR1EbEtIjIibppmn5zi57bhP5kkSc1p5BlvRBwJ3AIcAlwF3A68HDgLWBsRqzJzWx9VnQGcBDwG3AX0+x249wEbuiz/cZ/7S5JURFODqy6lCt0zM/OS1sKIuBA4G7gAOL2Pej4KvJ8quJ8H3Ntn+5OZed4gByxJ0mwY+lZz3dtdA0wCn+pYfS6wA1gXEQumqyszb83MH2TmE8MelyRJc1ETPd7j6/LazHyyfUVmPhIRN1MF80rgugba6+aAiPgr4FBgO7ApM32+K0mac5oI3qPqckuP9XdSBe8SRhe8LwH+rX1BRHwHWJeZ3+ungojY1GPV0iGPTZKkpzQxqnlhXW7vsb61/IAG2urmQmAV8Bxgf+BlwJVUYXx9RDx3RO1KkjSwsZ+5KjPP6Vi0EXhLRFwJvAl4D9UAr+nqWd5ted0TPmbY45QkCZrp8bZ6tAt7rG8tf7iBtgbxmbo8rnC7kiT11ESP9466XNJj/eK67PUMeFR+XpfTjqaW5qOp5lx2TmVp9jTR472hLtdExC71RcT+VM9fHwVKjzJeWZf3FG5XkqSehg7ezLwbuBZYRDXzVLv1VD3OyzNzR2thRCyNiKFHC0fEiyNi727LqSbtAPj8sO1IktSUpgZXvYNqyshPRMRqYAI4luod3y1Us1G1m6jLXe6FRcSrgLfXv+5Xl4sjYkNrm8w8tW2XdwMnRsQ3gPuBx6le/1kL7Al8FviPIT6XJEmNaiR4M/PuiFgBnE8Veq8DfgJcDKzPzIf6rOoFwCkdyw7pWHZq239/GXgW8GLgBGAfYBvwNeCzmfmVgT6IJEkj1tjrRJl5P3Ban9t2HfWRmRvo/mUHver5MlX4SpI0Fsb+PV5Jg5tqxLOk0Wrs+3glSdL0DF5JkgoyeCVJKsjglSSpIINXkqSCDF5JkgoyeCVJKsj3eKUx5vu40vixxytJUkEGryRJBRm8kiQVZPBKklSQg6skNWaqwV4bMwseiTR3GbySijCUpYq3miVJKsjglSSpIINXkqSCDF5JkgoyeCVJKsjglSSpIINXkqSCDF5JkgoyeCVJKsjglSSpIINXkqSCnKtZmuOmmuNY0vixxytJUkEGryRJBRm8kiQVZPBKklSQg6skzbqpBpBtzCx4JNLo2eOVJKkge7zSHOArQ9Luwx6vJEkFGbySJBVk8EqSVJDBK0lSQQavJEkFGbySJBVk8EqSVJDBK0lSQQavJEkFGbySJBVk8EqSVJBzNUua0/zmIs039nglSSrI4JUkqSCDV5KkggxeSZIKMnglSSrI4JUkqSCDV5KkggxeSZIKMnglSSrI4JUkqSCDV5KkggxeSZIKMnglSSrIbyeSCpnqW3Yk7T7s8UqSVJDBK0lSQY0Fb0QcHhGXRcTWiHg8IiYj4qKIOHCAOl4TER+LiOsiYltEZETc1Md+fxARX4iIByLisYi4IyLWR8S+w30qSZKa1cgz3og4ErgFOAS4CrgdeDlwFrA2IlZl5rY+qjoDOAl4DLgLOKiPto8Frgf2Bq4E7gdOAD4IrI6I1Zn5+MAfSpKkEWiqx3spVeiemZknZ+bfZ+YJwMeBo4AL+qzno8CLgP2AE6fbOCL2BD4HPBN4c2b+eWa+FzgW+CKwCjh70A8jSdKoDB28dW93DTAJfKpj9bnADmBdRCyYrq7MvDUzf5CZT/TZ/B8Dy4AbM/MrbfU8Cfxd/evpEQ4nlSTNDU30eI+vy2vrwHtKZj4C3EzVI13ZQFudTqjLqztXZOY9wBbg94EjRtC2JEkDayJ4j6rLLT3W31mXSxpoay61LUnSwJoYXLWwLrf3WN9afkADbY2s7YjY1GPV0gGPSZKknnyPV5Kkgpro8bZ6lQt7rG8tf7iBtkbWdmYu77a87gkfM/CRSZLURRM93jvqstdz1MV12es57Li2LUnSwJoI3hvqck1E7FJfROxP9S7to8BtDbTV6fq6XNu5IiKOoArk+4B7RtC2JEkDGzp4M/Nu4FpgEdXMU+3WAwuAyzNzR2thRCyNiCYGLX0dmACOi4g3tNW/B9VkHACfycxsoC1JkobW1NcCvoNqyshPRMRqqjA8luod3y3A+zu2n6jLXSa2iIhXAW+vf92vLhdHxIbWNpl5att/PxERp1H1fK+MiCuBHwGrgRVU7xB/fMjPJklSYxoJ3sy8OyJWAOdT3fZ9HfAT4GJgfWY+1GdVLwBO6Vh2SMeyUzva/mZEvIyqd70G2J/q9vL5wEecp1mSNJc01eMlM+8HTutz265TOGbmBmDDDNr+IfCWQfeTJKk03+OVJKmgxnq8klTaiim+/2SjYyo1R9njlSSpIINXkqSCDF5JkgoyeCVJKsjglSSpIEc1Sw2aapStJIE9XkmSijJ4JUkqyOCVJKkgg1eSpIIMXkmSCjJ4JUkqyOCVJKkg3+OVBuS7upKGYY9XkqSCDF5JkgoyeCVJKsjglSSpIINXkqSCDF5JkgoyeCVJKsjglSSpIINXkqSCDF5JkgoyeCVJKsi5mqUunI95/E31v+HGzIJHIu3KHq8kSQUZvJIkFWTwSpJUkMErSVJBBq8kSQUZvJIkFWTwSpJUkMErSVJBBq8kSQUZvJIkFWTwSpJUkMErSVJBBq8kSQUZvJIkFWTwSpJUkMErSVJBBq8kSQXtNdsHIEmlrYjouW5jZsEj0e7IHq8kSQUZvJIkFWTwSpJUkMErSVJBBq8kSQUZvJIkFWTwSpJUkMErSVJBBq8kSQUZvJIkFWTwSpJUkHM1S1Ib53HWqNnjlSSpIINXkqSCDF5JkgoyeCVJKsjBVdotTTWARpJGqbEeb0QcHhGXRcTWiHg8IiYj4qKIOHDAeg6q95us69la13t4j+0nIyJ7/Py0mU8nSVIzGunxRsSRwC3AIcBVwO3Ay4GzgLURsSozt/VRz8F1PUuA64ErgKXAacDrI+IVmXlPl123Axd1Wf6rwT+NJEmj09St5kupQvfMzLyktTAiLgTOBi4ATu+jng9The6FmXlOWz1nAhfX7aztst/DmXnejI9ekqRChr7VXPd21wCTwKc6Vp8L7ADWRcSCaerZD1hXb39ex+pPAvcBr42II4Y9ZkmSZksTz3iPr8trM/PJ9hWZ+QhwM/BMYOU09awE9gVurvdrr+dJ4JqO9to9IyL+IiLeFxFnRcTxEbHnoB9EkqRRa+JW81F1uaXH+jupesRLgOuGrIe6nk6HApd3LLs3Ik7LzK9P0eZTImJTj1VL+9lfkqR+NNHjXViX23usby0/YET1fA5YTRW+C4CjgX8BFgFfi4iXTNOuJEnFjP17vJm5vmPR94HTI+JXwDlUz4vf2Ec9y7str3vCxwx5mJIkAc0Eb6snurDH+tbyhwvV0/IZquA9rs/tJWlK00284rcXqR9N3Gq+oy67PXsFWFyXvZ7dNl1Py8/rcsrR1JIkldRE8N5Ql2siYpf6ImJ/YBXwKHDbNPXcBuwEVtX7tdezB9UArfb2ptMaRd1twg1JkmbF0MGbmXcD11INZjqjY/V6qh7n5Zm5o7UwIpZGxC6jhTPzV1Qjkxfw9Pd431nXf037zFURsazb+8ERsYjq3V+Azw/6mSRJGpWmBle9g2qqx09ExGpgAjiW6p3bLcD7O7afqMvOBybvA14NvDsiXgp8C1gGnAQ8wNOD/c+AcyLiRqoJNh4BjgReD+wDfBX45+E+miRJzWkkeDPz7ohYAZxPNaXj64CfUE3zuD4zH+qznm0R8QqqGa9OBv4I2Eb1ytAHM/PHHbvcQPX+7x9S3dJeQDX46iaq3vPlmY52kCTNHY29TpSZ91N9mUE/2/YcGpiZD1J9ucJZfdTzdaCvCTIkSZoLGvtaQEmSND2DV5KkggxeSZIKMnglSSrI4JUkqSCDV5KkggxeSZIKMnglSSpo7L+PV+pluq9wk6TZYI9XkqSC7PFKUkOmusuy0WnjVbPHK0lSQQavJEkFGbySJBVk8EqSVJDBK0lSQQavJEkF+TqRxpqTZEgaNwavJBXgO75q8VazJEkFGbySJBVk8EqSVJDBK0lSQQavJEkFGbySJBVk8EqSVJDBK0lSQQavJEkFGbySJBVk8EqSVJDBK0lSQQavJEkF+e1EmvP86j/Nd35z0e7FHq8kSQUZvJIkFWTwSpJUkMErSVJBBq8kSQUZvJIkFWTwSpJUkO/xStIc5ju+8489XkmSCjJ4JUkqyOCVJKkgn/FqTnA+Zkm7C3u8kiQVZI9XksaUI57Hkz1eSZIKMnglSSrI4JUkqSCDV5KkghxcpWJ8ZUiS7PFKklSUwStJUkEGryRJBfmMV5J2M068Mbvs8UqSVJA9Xkmah3yLYO6yxytJUkH2eNWotc99Lr8z2wchSXOYwauBdbuF9ZVZOA5JGkcGryTpKY54Hj2f8UqSVFBjwRsRh0fEZRGxNSIej4jJiLgoIg4csJ6D6v0m63q21vUePuq2JUm9rYiY8kf9aeRWc0QcCdwCHAJcBdwOvBw4C1gbEasyc1sf9Rxc17MEuB64AlgKnAa8PiJekZn3jKJtSZJKaOoZ76VUwXdmZl7SWhgRFwJnAxcAp/dRz4epQvfCzDynrZ4zgYvrdtaOqG1J0hBm2uvd3Z4dD32rue5xrgEmgU91rD4X2AGsi4gF09SzH7Cu3v68jtWfBO4DXhsRRzTdtiRJpTTR4z2+Lq/NzCfbV2TmIxFxM1U4rgSum6KelcC+dT2PdNTzZERcA/x13V7rdnNTbc9LPnORNA5mOpJ6XEdgNxG8R9Xllh7r76QKvyVMHX791ENdT9NtExGbeqx6ycTEBMuXL59q977cvnlzz3VLjzlmRvvNFa37/3fiUHlJzVkwww7EVPtN9fe2XxMTEwCLZrJvE8G7sC6391jfWn7ACOppqu2pPLFz587tmzdvnhyijmltHoNwncr36vKx7quX1uXtJY5lN+Z5LsdzXU7j57qhv7eLgF/OZEcn0Khl5vBdWnXVupvgOR4tz3M5nuty5uO5buKuYKtXubDH+tbyh0dQT1NtS5JURBPBe0ddLumxfnFd9noOO0w9TbUtSVIRTQTvDXW5JiJ2qS8i9gdWAY8Ct01Tz23ATmBVvV97PXtQDZJqb6/JtiVJKmLo4M3Mu4FrqR40n9Gxej2wALg8M3e0FkbE0ohY2r5hZv4KuLze/ryOet5Z139N+8xVM2lbkqTZFNnAu05dpm2cAI6les92C/DK9mkbIyIBMjM66umcMvJbwDLgJOCBup67h2lb5c3HwRFzkee5HM91OfPxXDcSvAAR8TzgfKpXOg8GfgJ8CVifmQ91bNs1eOt1B1HNOnUy8HvANuBrwAcz88fDti1J0mxqLHglSdL0nGRIkqSCDF5JkgoyeCVJKsjglSSpIINXkqSCDF5JkgoyeDUjEXFwRLw9Ir4UEXdFxM6I2B4RN0XE2zqn8Gzb75UR8dWIeLDe57sR8a6I2LP0ZxgXg57riFgUETnFzxWz9VnGQUR8NCKui4j763P9YER8OyLOrSf56baP1/UMDHKu59N17Xu8mpGIOB34NNVkJTcAPwJ+F/gTqm+F+iLwlmy7wCLipHr5Y8B/Ag8CJwJHAVdm5ltKfoZxMei5johFwL3Ad4Avd6ny+5l55cgPfExFxP8Bm4EfUs2YtwBYCawAtgIrM/P+tu29rmdokHM9r67rzPTHn4F/gBOo/rjs0bH8UKpgSOBNbcufRfV/rMeBFW3L96Ga8jOBt87255qLPzM414vqZRtm+9jH8QfYp8fyC+rzemnbMq/rcud63lzX3mrWjGTm9Zn5X5n5ZMfynwKfqX99dduqNwPPAa7IzI1t2z8G/EP969+O7ojH1wzOtYZQX5PdfKEuF7ct87oewoDnet7Ya7YPQPPSr+vyN23LTqjLq7tsfyPV1ze+MiKekZmPj/Lg5plu57rlsIj4G6r5y7cBt2bmd4sd2fxzYl22n0Ov69Hodq5bxv66NnjVqIjYC/jL+tf2P0ZH1eWWzn0y8zcRcS/wQuAIqm+Y0jSmONctr6l/2vf5H+CUzPzRaI9u/EXEe4D9qJ6jrwBeRRUEH2nbzOu6AX2e65axv64NXjXtI8CLgK9m5jVtyxfW5fYe+7WWHzCi45qPep3rR4F/pBqA0vr+6hdTfc/18cB1EfHS9Huqp/MeqkFsLVcDp2bmz9uWeV03o59zPW+ua5/xqjERcSZwDnA7sG6WD2dem+pcZ+YDmfnBzNycmQ/XPzcCa4BvAi8A3l78oMdMZh6a1VeXHko1gvwI4NsRcczsHtn808+5nk/XtcGrRkTEO4GLqV4LOD4zH+zYpPUv/4V011r+cPNHN7/0ca67yszfAP9a/3rciA5v3snMn2Xml6j+wB8M/Hvbaq/rBk1zrnvtM3bXtcGroUXEu4BLgO9TBcFPu2x2R10u6bL/XsDzqQYI3dO5Xr/V57meSuvW3YImj2t3kJn3Uf1j54UR8ex6sdf1CPQ411MZq+va4NVQIuK9wMeB/6UKggd6bHp9Xa7tsu444JnALY787G2Acz2VlXVpEMzMYXX5RF16XY9O57meylhd1wavZiwiPkA1wGcTsDozfzHF5lcCvwDeGhEr2urYB/hQ/eunR3Ws426Qcx0Rx3SbsjMiVgNn179+fiQHOuYiYklEPO22cUTsEREXAIdQBelD9Sqv6xka9FzPp+vaKSM1IxFxCrCB6l+jl9B9VOdkZm5o2+dkqj9UjwFXUE2t9wbqqfWAP00vyKcZ9FzXr1Ysppo56cf1+hfz23dOP5CZH+qsQE/dyv8n4Caq6Qm3UY22/WOqAT8/pfqHzw/b9jkZr+uBDXqu59N1bfBqRiLiPODcaTb7ema+umO/VcD7gVdQTat3F3AZ8InM7OeW0m5n0HMdEW8D3kj1qtGzgb2BnwG3Ap/MzG+M7GDHXES8CDid6j3Sw6leA9pB9Z7uf1Ndp08bzOZ1PbhBz/V8uq4NXkmSCvIZryRJBRm8kiQVZPBKklSQwStJUkEGryRJBRm8kiQVZPBKklSQwStJUkEGryRJBRm8kiQVZPBKklSQwStJUkEGryRJBRm8kiQVZPBKklSQwStJUkEGryRJBf0/Tx5jRCYeraYAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 504x335.952 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "carbon confounded by volume, n=33, p=0.12121212121212122\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAcsAAAE7CAYAAACykKpeAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Il7ecAAAACXBIWXMAABYlAAAWJQFJUiTwAAAWd0lEQVR4nO3de7BdZXnH8e9DQrlVriU6qG2AAkEN0AQFyZTrwAQdARUq04pYhcErgtDR4dagOPpHR1BsoXKLhZkCQsVRKTpDCOFi6xDKbQwgQqBUMEAgECBI4Okfax08Oex93nNZ+6x9Tr6fmTXr7HV7n/NmZ//O2nutd0dmIkmSutug7QIkSep3hqUkSQWGpSRJBYalJEkFhqUkSQWGpSRJBYalJEkFhqUkSQWGpSRJBYalJEkFhqUkSQWGpSRJBdPbLqAXIuIRYHNgeculSJL6x0zg+czcfrQ7TsmwBDbfZJNNtt511123bruQ8Xj1zju5F5g9ezYbbrhh2+VI0qS2bNkyXn755THtO+6wjIhtgA8DHwRmA28H/gDcC1wGXJaZrw/afibwyDCHvCozjx5nWct33XXXrZcuXTrOw7TrdxG8HbjhhhvYbrvt2i5Hkia1uXPncueddy4fy75NnFkeBVwAPAHcBDwGvBX4CHAxcGhEHJVv/uLMu4HrOhzvvgZqkiSpMU2E5YPAYcDPhpxBngb8CvgoVXBeO2S/uzJzQQPtS5LUU+O+GjYzF2XmTwYHZb38SeDC+uH+421HkqS29PoCn1fr+doO67aLiBOAbYBngF9m5j09rkeSpFHrWVhGxHTgE/XDGzpscnA9Dd5nMXBsZj42wja6XcEza4RlSpJU1MtBCb4FvAe4PjN/Pmj5S8DXgbnAVvW0H9XFQfsDN0bEZj2sS5KkUenJmWVEnAicAtwPHDN4XWauAM4assuSiDgEuBXYCzgO+E6pncyc26X9pcCc0VcuSdKbNX5mGRFfoAq6XwMHZObKkeyXmWupbjUB2LfpuiRJGqtGwzIiTgLOp7pX8oD6itjReKqe+zasJKlvNBaWEfEV4FzgLqqgXDGGw+xdzx9uqi5Jksarkc8sI+JM4GvAUuCQ4d56jYg5VAMSvD5k+UHAyfXDK5qoqwkzZszoum7FirH8PaD13WR6Tk2mWqVeamJs2GOpgvI14BbgxIgYutnyzFxY//xtYKeIuB14vF62G3Bg/fOZmXn7eOuSJKkpTZxZDnzVyTTgpC7b3AwsrH++nGrg9fcChwIbAr8Hrga+l5m3NFCTJEmNGXdY1uO7LhjF9pcAl4y3XUmSJkovByWQJGlKMCwlSSowLCVJKjAsJUkqMCwlSSowLCVJKjAsJUkqMCwlSSowLCVJKjAsJUkqMCwlSSowLCVJKjAsJUkqMCwlSSowLCVJKjAsJUkqMCwlSSowLCVJKjAsJUkqMCwlSSowLCVJKjAsJUkqMCwlSSowLCVJKjAsJUkqMCwlSSowLCVJKjAsJUkqMCwlSSowLCVJKjAsJUkqMCwlSSowLCVJKjAsJUkqMCwlSSoYd1hGxDYRcVxE/CgiHoqIlyNiVUTcGhGfjoiObUTEPhFxfUSsrPe5JyJOiohp461JkqQmTW/gGEcBFwBPADcBjwFvBT4CXAwcGhFHZWYO7BARhwPXAmuAq4CVwIeAc4F59TElSeoLTYTlg8BhwM8y8/WBhRFxGvAr4KNUwXltvXxz4CLgNWD/zLyjXn4msAg4MiKOzswrG6hNkqRxG/fbsJm5KDN/Mjgo6+VPAhfWD/cftOpIYFvgyoGgrLdfA5xRP/zseOuSJKkpvb7A59V6vnbQsgPr+Q0dtl8CvATsExEb9bIwSZJGqom3YTuKiOnAJ+qHg4Nxl3r+4NB9MnNtRDwCvBvYAVhWaGNpl1WzRletJEnd9fLM8lvAe4DrM/Png5ZvUc9XddlvYPmWPapLkqRR6cmZZUScCJwC3A8c04s2ADJzbpf2lwJzetWuJGn90viZZUR8AfgO8GvggMxcOWSTgTPHLehsYPlzTdcmSdJYNBqWEXEScD5wH1VQPtlhswfq+c4d9p8ObE91QdDDTdYmSdJYNRaWEfEVqkEF7qIKyhVdNl1Uz+d3WLcvsClwe2a+0lRtkiSNRyNhWQ8o8C1gKXBQZj49zObXAE8DR0fEnoOOsTFwTv3wgibqkiSpCeO+wCcijgW+RjUizy3AiRExdLPlmbkQIDOfj4jjqUJzcURcSTXc3WFUt5VcQzUEniRJfaGJq2G3r+fTgJO6bHMzsHDgQWZeFxH7AadTDYe3MfAQ8GXgu4PHkZUkqW3jDsvMXAAsGMN+twEfGG/7kiT1mt9nKUlSgWEpSVKBYSlJUoFhKUlSgWEpSVKBYSlJUoFhKUlSgWEpSVKBYSlJUoFhKUlSgWEpSVKBYSlJUoFhKUlSgWEpSVKBYSlJUoFhKUlSgWEpSVKBYSlJUoFhKUlSgWEpSVKBYSlJUoFhKUlSgWEpSVKBYSlJUoFhKUlSgWEpSVKBYSlJUsH0tgtQ782YMWPY9StWrJigSiqlerqZ6DrbMFzfDPf7j3W/yWR9+B3VvzyzlCSpwLCUJKnAsJQkqcCwlCSpwLCUJKnAsJQkqaCRsIyIIyPi/Ii4JSKej4iMiCu6bDuzXt9turKJmiRJakpT91meAewOrAYeB2aNYJ+7ges6LL+voZokSWpEU2F5MlVIPgTsB9w0gn3uyswFDbUvSVLPNBKWmflGOEZEE4eUJKlvtDnc3XYRcQKwDfAM8MvMvKfFeiRJ6qjNsDy4nt4QEYuBYzPzsZEcICKWdlk1ks9MJUkakTZuHXkJ+DowF9iqngY+59wfuDEiNmuhLkmSOprwM8vMXAGcNWTxkog4BLgV2As4DvjOCI41t9Py+oxzzjhLlSQJ6KNBCTJzLXBx/XDfNmuRJGmwvgnL2lP13LdhJUl9o9/Ccu96/nCrVUiSNMiEh2VEzImIN7UbEQdRDW4A0HGoPEmS2tDIBT4RcQRwRP3wbfX8/RGxsP756cw8tf7528BOEXE71ag/ALsBB9Y/n5mZtzdRlyRJTWjqatg9gGOHLNuhngAeBQbC8nLgw8B7gUOBDYHfA1cD38vMWxqqSZKkRjQ13N0CYMEIt70EuKSJdiVJmgj9doGPJEl9x7CUJKnAsJQkqcCwlCSpwLCUJKnAsJQkqcCwlCSpwLCUJKnAsJQkqcCwlCSpwLCUJKnAsJQkqcCwlCSpwLCUJKnAsJQkqcCwlCSpwLCUJKnAsJQkqcCwlCSpwLCUJKnAsJQkqcCwlCSpwLCUJKnAsJQkqcCwlCSpwLCUJKnAsJQkqcCwlCSpwLCUJKnAsJQkqcCwlCSpwLCUJKnAsJQkqcCwlCSpoJGwjIgjI+L8iLglIp6PiIyIKwr77BMR10fEyoh4OSLuiYiTImJaEzVJktSU6Q0d5wxgd2A18Dgwa7iNI+Jw4FpgDXAVsBL4EHAuMA84qqG6JEkat6behj0Z2BnYHPjscBtGxObARcBrwP6Z+enM/AdgD+CXwJERcXRDdUmSNG6NhGVm3pSZv8nMHMHmRwLbAldm5h2DjrGG6gwVCoErSdJEauMCnwPr+Q0d1i0BXgL2iYiNJq4kSZK6a+ozy9HYpZ4/OHRFZq6NiEeAdwM7AMuGO1BELO2yatjPTCVJGo02ziy3qOeruqwfWL5l70uRJKmsjTPLxmTm3E7L6zPOORNcjiRpimrjzHLgzHGLLusHlj/X+1IkSSprIywfqOc7D10REdOB7YG1wMMTWZQkSd20EZaL6vn8Duv2BTYFbs/MVyauJEmSumsjLK8BngaOjog9BxZGxMbAOfXDC1qoS5Kkjhq5wCcijgCOqB++rZ6/PyIW1j8/nZmnAmTm8xFxPFVoLo6IK6mGuzuM6raSa6iGwJMkqS80dTXsHsCxQ5btUE8AjwKnDqzIzOsiYj/gdOCjwMbAQ8CXge+OcCQgSZImRCNhmZkLgAWj3Oc24ANNtC9JUi/5fZaSJBUYlpIkFRiWkiQVGJaSJBUYlpIkFRiWkiQVGJaSJBUYlpIkFRiWkiQVTOovf56KZsyY8cbPd9Xz2bNnM23atFbqGc7gWodasWLFBFYyfC3jMdbfo1f1jEU//TtB/9UzVlPl9+gn/dynnllKklRgWEqSVGBYSpJUYFhKklRgWEqSVGBYSpJUYFhKklRgWEqSVGBYSpJUYFhKklRgWEqSVGBYSpJUYFhKklRgWEqSVGBYSpJUYFhKklRgWEqSVGBYSpJUYFhKklRgWEqSVGBYSpJUYFhKklRgWEqSVGBYSpJU0FpYRsTyiMgu05Nt1SVJ0lDTW25/FXBeh+WrJ7gOSZK6ajssn8vMBS3XIEnSsPzMUpKkgrbPLDeKiI8Dfw68CNwDLMnM19otS5KkP2o7LN8GXD5k2SMR8feZeXNp54hY2mXVrHFXJklSrc23YS8DDqIKzM2A2cC/AjOB/4yI3dsrTZKkP2rtzDIzzx6y6D7gMxGxGjgFWAB8uHCMuZ2W12eccxooU5KkvrzA58J6vm+rVUiSVOvHsHyqnm/WahWSJNX6MSz3rucPt1qFJEm1VsIyInaNiDedOUbETOB79cMrJrQoSZK6aOsCn48Bp0TEEuBR4AVgR+CDwMbA9cA/tVSbJEnraCssbwJ2Af4KmEf1+eRzwK1U911enpnZUm2SJK2jlbCsBxwoDjogSVI/6McLfCRJ6iuGpSRJBYalJEkFhqUkSQWGpSRJBYalJEkFhqUkSQWGpSRJBYalJEkFhqUkSQWGpSRJBYalJEkFhqUkSQWGpSRJBYalJEkFhqUkSQWGpSRJBYalJEkFhqUkSQWGpSRJBYalJEkFhqUkSQWGpSRJBYalJEkFhqUkSQWGpSRJBYalJEkFhqUkSQWGpSRJBYalJEkFhqUkSQWGpSRJBYalJEkFhqUkSQWGpSRJBa2GZUS8IyIujYjfRcQrEbE8Is6LiK3arEuSpMGmt9VwROwI3A7MAH4M3A+8D/gSMD8i5mXmM23VJ0nSgDbPLP+FKihPzMwjMvOrmXkgcC6wC/CNFmuTJOkNrYRlfVZ5CLAc+Ochq/8ReBE4JiI2m+DSJEl6k7bOLA+o57/IzNcHr8jMF4DbgE2BvSe6MEmShmrrM8td6vmDXdb/hurMc2fgxm4HiYilXVbtvmzZMubOnTv2CmvPPvts13VNHH+49ubX81WrVhERjbc1YKy/x1j7Zrj9htOLY46nzeH0op6J7tPScfvpedOL/4sl/VbPVNDrPl22bBnAzLHsG5k57gJG3WjE94HjgeMz8+IO678BnAaclpnfHOY43cLyPcBqqrd52zSrnt/fahXrJ/u+XfZ/u+z/zmYCz2fm9qPdsbWrYZuQmX3959tAmPd7nVORfd8u+79d9n/z2vrMclU936LL+oHlz/W+FEmShtdWWD5Qz3fusn6net7tM01JkiZMW2F5Uz0/JCLWqSEi3gLMA14C/muiC5MkaahWwjIzfwv8gurD1s8PWX02sBlweWa+OMGlSZL0Jm1e4PM5quHuvhsRBwHLgL2o7sF8EDi9xdokSXpDK7eOvNF4xDuBr1HdUrgN8ATwI+DszOzNjXOSJI1Sq2EpSdJk4PdZSpJUYFhKklRgWEqSVGBYSpJUYFhKklRgWEqSVGBYdhAR74iISyPidxHxSkQsj4jzImKrcRxz34h4LSIyIs4ZZrt9IuL6iFgZES9HxD0RcVJETBtr25NNG/0fETPrdd2mK8f3W00OTfR9RCwu9OXGXfZ7V0RcHRErImJNRDwQEWdHxCbN/Yb9ra3+L2zvsKNM8q/o6oWI2JFqZKEZwI+pvg/ufcCXgPkRMS8znxnlMd8C/IBqvNs/HWa7w4FrgTXAVcBK4EPAuVTj5R412t9nsmmz/2t3A9d1WH7faNqcjHrQ92d3Wb62Q9t7AYuADYFrgP8FDgTOAg6KiIMy85VRtD3ptNn/tUeBhR2WPz6KNqeuzHQaNAE/BxL44pDl366XXziGY15KFXyn1cc4p8M2mwMrgFeAPQct35jqP1ACR7fdP1O4/2fW6xa23QeTve+BxdVLy4jbnQb8um7jsEHLN6AKzgS+2nb/TNX+r/dJYHHbfdDPU+sF9NME7Fg/aR4BNhiy7i3AauBFYLNRHPPw+pgfBz45zIv1p+p1P+iw7sB63c1t99EU7v/1Oiyb7PsxhGXX5zewQ71uOfWIY1NxarP/630My8LkZ5brOqCe/yIzXx+8IjNfAG4DNgX2HsnBImIGcBFwXWZeUdj8wHp+Q4d1S6jeQtwnIjYaSduTVJv9P2C7iDghIk6r57uNcL/JrtG+B4iIj0XEVyPiyxFx6DDP3a7P/cx8mOqLFf6CKjinqjb7f8CWEfGp+rn/+YgYcVvrAz+zXNcu9bzbl07/BjiE6kurbxzB8S6ieivpM+NpOzPXRsQjwLupXjCWjeB4k1Gb/T/g4Hp6Q0QsBo7NzMdGcZzJpum+Bxh6UdSKiPh8Zl4zhrZ3rqffjrDtyabN/h+wO3DJ4AURcTdwTGbeO8I2pyzPLNe1RT1f1WX9wPItSweKiE8BhwGfy8zfT2Tbk1ib/f8S8HVgLrBVPe1H9UXl+wM3RsRmIzjOZNXk8+/HVBemvQPYBJgFfLPe96qImN/DtierNvsfqs9F5wHbUr3t+16qz4t3BxZFxNtH0O6UZlj2QETMBM4DfpiZV7dbzfpnLP2fmSsy86zMvDMzn6unJVR/zf838JfAcb2qeSrJzHMz86eZ+X+ZuSYzH8jM04BTqF5zvtlyiVPaWPo/M0/JzNsz8+nMXJ2Zd2TmUVRX5/8ZcOrE/hb9x7Bc18Bfb1t0WT+w/LnCcS4FXqb6guuJbnsya7P/O8rMtcDF9cN9x3u8PjYRz7+LqW5b2KO+nWci2+53bfb/cC6s51P5uT8ihuW6HqjnO3dZv1M97/a5woA5VPdKPTX45l7gsnr96fWy60bSdkRMB7aneqI/XGh7Mmuz/4fzVD2fym/DNtX3XWXmGuCF+uHgvux525NAm/0/nPXhuT8iXuCzrpvq+SERscHgq9Lqv8TmUX22VRrR4t+orlwbaieqv9DuApYC/zNo3SLg74D5wL8P2W/f+nhLcmrfmN1m/w9n4KrAqfyHSlN931VE7EL1WfALwNODVi0CTqd67n9zyD47UAXIo9j/ver/4awPz/2RafvelX6bGOWNwVQfns8a4bE/yfCDEjyFgxK01f9zGHJ/W738IKoRlRLYp+3+6fe+p3oHZOsOx9520PP4+0PWDTcowQ9xUIJe9/9uwIYd9tmNKlQT+Nu2+6ftKepOUa3DkFPLgL2o7oN6kOoF85lB21d39GbGCI79Saq3Ar+RmWd0WH8E1RVoa6gu+15JdUXnLvXyv8kp/g/WVv/Xt4fsVLc9MLzXbvzxHsAzM7PrmL5TQRN9X/fxhcCtVGcjK4E/Bz5A9bnbHcDBmfnckLaHDnf3GNUfKntS3WO4Pg53NyH9HxELqa6evYVqmMFXqIJ4PtUfMhcBJ0z1156ittO6HyfgnVQvqk8Af6B6C+g8YKsO2yYjHC2DYc5sBm0zD7geeJbqIpV7gZOBaW33y1Tuf+DTwE+pRopZTfWC8RjVGL1/3XafTJa+B2ZTjS96L/AM8CrVC/YtwBeBPxmm7XdRnUk+Xff/g1Tjm27Sdr9M5f4HjgD+A3gIeL5u9wngJww601/fJ88sJUkq8GpYSZIKDEtJkgoMS0mSCgxLSZIKDEtJkgoMS0mSCgxLSZIKDEtJkgoMS0mSCgxLSZIKDEtJkgoMS0mSCgxLSZIKDEtJkgoMS0mSCgxLSZIKDEtJkgr+H8XsZmhrRLuuAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 504x335.952 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# plots\n",
    "# histogram of nitrogen conservation\n",
    "plot_code_histograms(res, \"nitrogen\")\n",
    "# carbon conservation\n",
    "plot_code_histograms(res, \"carbon\")\n",
    "# molecular volume\n",
    "plot_code_histograms(res, \"volume\")\n",
    "# carbon conservation confounded by volume\n",
    "plot_code_histograms(res, \"carbon\", \"volume\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
